// @effect/platform/dist/esm/HttpClientResponse.js
import * as internal from "./internal/httpClientResponse.js";
export {
/**
 * @since 1.0.0
 * @category schema
 */
schemaBodyJson,
/**
 * @since 1.0.0
 * @category schema
 */
schemaBodyUrlParams,
/**
 * @since 1.0.0
 * @category schema
 */
schemaHeaders } from "./HttpIncomingMessage.js";
/**
 * @since 1.0.0
 * @category type ids
 */
export const TypeId = internal.TypeId;
/**
 * @since 1.0.0
 * @category constructors
 */
export const fromWeb = internal.fromWeb;
/**
 * @since 1.0.0
 * @category schema
 */
export const schemaJson = internal.schemaJson;
/**
 * @since 1.0.0
 * @category schema
 */
export const schemaNoBody = internal.schemaNoBody;
/**
 * @since 1.0.0
 * @category accessors
 */
export const stream = internal.stream;
/**
 * @since 1.0.0
 * @category pattern matching
 */
export const matchStatus = internal.matchStatus;
/**
 * @since 1.0.0
 * @category filters
 */
export const filterStatus = internal.filterStatus;
/**
 * @since 1.0.0
 * @category filters
 */
export const filterStatusOk = internal.filterStatusOk;
