// @evolution-sdk/evolution/dist/plutus/Credential.js
import * as Data from "../Data.js";
import * as TSchema from "../TSchema.js";
/**
 * Verification Key Hash (28 bytes)
 */
export const VerificationKeyHash = TSchema.ByteArray;
/**
 * Script Hash (28 bytes)
 */
export const ScriptHash = TSchema.ByteArray;
/**
 * Credential - Either a verification key hash or script hash
 */
export const Credential = /*#__PURE__*/TSchema.Variant({
  VerificationKey: {
    hash: VerificationKeyHash
  },
  Script: {
    hash: ScriptHash
  }
});
/**
 * Payment Credential - Used for payment addresses
 */
export const PaymentCredential = /*#__PURE__*/TSchema.Variant({
  VerificationKey: {
    hash: VerificationKeyHash
  },
  Script: {
    hash: ScriptHash
  }
});
/**
 * Stake Credential - Either inline credential or pointer
 */
export const StakeCredential = /*#__PURE__*/TSchema.Variant({
  Inline: {
    credential: Credential
  },
  Pointer: {
    slot_number: TSchema.Integer,
    transaction_index: TSchema.Integer,
    certificate_index: TSchema.Integer
  }
});
// Export codec objects with all conversion functions
export const CredentialCodec = /*#__PURE__*/Data.withSchema(Credential);
export const PaymentCredentialCodec = /*#__PURE__*/Data.withSchema(PaymentCredential);
export const StakeCredentialCodec = /*#__PURE__*/Data.withSchema(StakeCredential);
export const VerificationKeyHashCodec = /*#__PURE__*/Data.withSchema(VerificationKeyHash);
export const ScriptHashCodec = /*#__PURE__*/Data.withSchema(ScriptHash);
