// @evolution-sdk/evolution/dist/sdk/provider/Koios.js
import { Effect } from "effect";
import * as KoiosEffect from "./internal/KoiosEffect.js";
/**
 * Koios provider for Cardano blockchain data access.
 * Provides support for interacting with the Koios API across multiple environments.
 * Supports optional bearer token authentication.
 *
 * @since 2.0.0
 * @category constructors
 */
export class Koios {
  baseUrl;
  token;
  Effect;
  constructor(baseUrl, token) {
    this.baseUrl = baseUrl;
    this.token = token;
    this.Effect = {
      getProtocolParameters: () => KoiosEffect.getProtocolParameters(this.baseUrl, this.token),
      getUtxos: KoiosEffect.getUtxos(this.baseUrl, this.token),
      getUtxosWithUnit: KoiosEffect.getUtxosWithUnit(this.baseUrl, this.token),
      getUtxoByUnit: KoiosEffect.getUtxoByUnit(this.baseUrl, this.token),
      getUtxosByOutRef: KoiosEffect.getUtxosByOutRef(this.baseUrl, this.token),
      getDelegation: KoiosEffect.getDelegation(this.baseUrl, this.token),
      getDatum: KoiosEffect.getDatum(this.baseUrl, this.token),
      awaitTx: KoiosEffect.awaitTx(this.baseUrl, this.token),
      submitTx: KoiosEffect.submitTx(this.baseUrl, this.token),
      evaluateTx: KoiosEffect.evaluateTx(this.baseUrl, this.token)
    };
  }
  getProtocolParameters = () => Effect.runPromise(this.Effect.getProtocolParameters());
  getUtxos = addressOrCredential => Effect.runPromise(this.Effect.getUtxos(addressOrCredential));
  getUtxosWithUnit = (addressOrCredential, unit) => Effect.runPromise(this.Effect.getUtxosWithUnit(addressOrCredential, unit));
  getUtxoByUnit = unit => Effect.runPromise(this.Effect.getUtxoByUnit(unit));
  getUtxosByOutRef = outRefs => Effect.runPromise(this.Effect.getUtxosByOutRef(outRefs));
  getDelegation = rewardAddress => Effect.runPromise(this.Effect.getDelegation(rewardAddress));
  getDatum = datumHash => Effect.runPromise(this.Effect.getDatum(datumHash));
  awaitTx = (txHash, checkInterval, timeout) => Effect.runPromise(this.Effect.awaitTx(txHash, checkInterval, timeout));
  submitTx = tx => Effect.runPromise(this.Effect.submitTx(tx));
  evaluateTx = (tx, additionalUTxOs) => Effect.runPromise(this.Effect.evaluateTx(tx, additionalUTxOs));
}
