// @evolution-sdk/evolution/dist/sdk/provider/Kupmios.js
import { Effect } from "effect";
import * as KupmiosEffects from "./internal/KupmiosEffects.js";
/**
 * Kupmios provider for Cardano blockchain data access.
 * Provides support for interacting with both Kupo and Ogmios APIs.
 * Supports custom headers for authentication with Demeter or other providers.
 *
 * @since 2.0.0
 * @category constructors
 */
export class KupmiosProvider {
  kupoUrl;
  ogmiosUrl;
  headers;
  Effect;
  constructor(kupoUrl, ogmiosUrl, headers) {
    this.kupoUrl = kupoUrl;
    this.ogmiosUrl = ogmiosUrl;
    this.headers = headers;
    this.Effect = {
      getProtocolParameters: () => KupmiosEffects.getProtocolParametersEffect(this.ogmiosUrl, this.headers),
      getUtxos: KupmiosEffects.getUtxosEffect(this.kupoUrl, this.headers),
      getUtxosWithUnit: KupmiosEffects.getUtxosWithUnitEffect(this.kupoUrl, this.headers),
      getUtxoByUnit: KupmiosEffects.getUtxoByUnitEffect(this.kupoUrl, this.headers),
      getUtxosByOutRef: KupmiosEffects.getUtxosByOutRefEffect(this.kupoUrl, this.headers),
      getDelegation: KupmiosEffects.getDelegationEffect(this.ogmiosUrl, this.headers),
      getDatum: KupmiosEffects.getDatumEffect(this.kupoUrl, this.headers),
      awaitTx: KupmiosEffects.awaitTxEffect(this.kupoUrl, this.headers),
      evaluateTx: KupmiosEffects.evaluateTxEffect(this.ogmiosUrl, this.headers),
      submitTx: KupmiosEffects.submitTxEffect(this.ogmiosUrl, this.headers)
    };
  }
  getProtocolParameters = () => Effect.runPromise(this.Effect.getProtocolParameters());
  getUtxos = addressOrCredential => Effect.runPromise(this.Effect.getUtxos(addressOrCredential));
  getUtxosWithUnit = (addressOrCredential, unit) => Effect.runPromise(this.Effect.getUtxosWithUnit(addressOrCredential, unit));
  getUtxoByUnit = unit => Effect.runPromise(this.Effect.getUtxoByUnit(unit));
  getUtxosByOutRef = outRefs => Effect.runPromise(this.Effect.getUtxosByOutRef(outRefs));
  getDelegation = rewardAddress => Effect.runPromise(this.Effect.getDelegation(rewardAddress));
  getDatum = datumHash => Effect.runPromise(this.Effect.getDatum(datumHash));
  awaitTx = (txHash, checkInterval, timeout) => Effect.runPromise(this.Effect.awaitTx(txHash, checkInterval, timeout));
  evaluateTx = (tx, additionalUTxOs) => Effect.runPromise(this.Effect.evaluateTx(tx, additionalUTxOs));
  submitTx = tx => Effect.runPromise(this.Effect.submitTx(tx));
}
