// @evolution-sdk/evolution/dist/Anchor.js
import { Equal, FastCheck, Hash, Inspectable, Schema } from "effect";
import * as Bytes from "./Bytes.js";
import * as Bytes32 from "./Bytes32.js";
import * as CBOR from "./CBOR.js";
import * as Url from "./Url.js";
/**
 * Schema for Anchor representing an anchor with URL and data hash.
 * ```
 * anchor = [anchor_url: url, anchor_data_hash: Bytes32]
 * ```
 *
 * @since 2.0.0
 * @category schemas
 */
export class Anchor extends /*#__PURE__*/Schema.TaggedClass()("Anchor", {
  anchorUrl: Url.Url,
  anchorDataHash: Bytes32.BytesFromHex
}) {
  toJSON() {
    return {
      _tag: "Anchor",
      anchorUrl: this.anchorUrl.href,
      anchorDataHash: Bytes.toHex(this.anchorDataHash)
    };
  }
  toString() {
    return Inspectable.format(this.toJSON());
  }
  [Inspectable.NodeInspectSymbol]() {
    return this.toJSON();
  }
  [Equal.symbol](that) {
    return that instanceof Anchor && Equal.equals(this.anchorUrl, that.anchorUrl) && Bytes.equals(this.anchorDataHash, that.anchorDataHash);
  }
  [Hash.symbol]() {
    return Hash.cached(this, Hash.combine(Hash.hash("Anchor"))(Hash.combine(Hash.hash(this.anchorUrl))(Hash.array(Array.from(this.anchorDataHash)))));
  }
}
export const CDDLSchema = /*#__PURE__*/Schema.Tuple(CBOR.Text,
// anchor_url: url
CBOR.ByteArray // anchor_data_hash: Bytes32
);
/**
 * CDDL schema for Anchor as tuple structure.
 * ```
 * anchor = [anchor_url: url, anchor_data_hash: Bytes32]
 * ```
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromCDDL = /*#__PURE__*/Schema.transform(CDDLSchema, /*#__PURE__*/Schema.typeSchema(Anchor), {
  strict: true,
  encode: toA => [toA.anchorUrl.href, toA.anchorDataHash],
  decode: ([anchorUrl, anchorDataHash]) => new Anchor({
    anchorUrl: new Url.Url({
      href: anchorUrl
    }),
    anchorDataHash
  })
});
/**
 * CBOR bytes transformation schema for Anchor.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromCBORBytes = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.compose(CBOR.FromBytes(options),
// Uint8Array → CBOR
FromCDDL // CBOR → Anchor
);
/**
 * CBOR hex transformation schema for Anchor.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromCBORHex = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.compose(Schema.Uint8ArrayFromHex,
// string → Uint8Array
FromCBORBytes(options) // Uint8Array → Anchor
);
/**
 * FastCheck arbitrary for Anchor instances.
 *
 * @since 2.0.0
 * @category arbitrary
 */
export const arbitrary = /*#__PURE__*/FastCheck.record({
  anchorUrl: Url.arbitrary,
  anchorDataHash: FastCheck.uint8Array({
    minLength: 32,
    maxLength: 32
  })
}).map(({
  anchorDataHash,
  anchorUrl
}) => new Anchor({
  anchorUrl,
  anchorDataHash
}, {
  disableValidation: true
})); // Disable validation since we already check length in FastCheck
// ============================================================================
// Parsing Functions
// ============================================================================
/**
 * Parse an Anchor from CBOR bytes.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromCBORBytes = (bytes, options) => Schema.decodeSync(FromCBORBytes(options))(bytes);
/**
 * Parse an Anchor from CBOR hex string.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromCBORHex = (hex, options) => Schema.decodeSync(FromCBORHex(options))(hex);
// ============================================================================
// Encoding Functions
// ============================================================================
/**
 * Convert an Anchor to CBOR bytes.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toCBORBytes = (anchor, options) => Schema.encodeSync(FromCBORBytes(options))(anchor);
/**
 * Convert an Anchor to CBOR hex string.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toCBORHex = (anchor, options) => Schema.encodeSync(FromCBORHex(options))(anchor);
