// @evolution-sdk/evolution/dist/BigInt.js
import { Schema } from "effect";
/**
 * BigUInt schema based on Conway CDDL specification
 *
 * CDDL: big_uint = #6.2(bounded_bytes)
 *
 * @since 2.0.0
 * @category schemas
 */
export const BigUIntSchema = /*#__PURE__*/Schema.TaggedStruct("BigUInt", {
  bytes: Schema.Uint8ArrayFromSelf
});
/**
 * BigNInt schema based on Conway CDDL specification
 *
 * CDDL: big_nint = #6.3(bounded_bytes)
 *
 * @since 2.0.0
 * @category schemas
 */
export const BigNIntSchema = /*#__PURE__*/Schema.TaggedStruct("BigNInt", {
  bytes: Schema.Uint8ArrayFromSelf
});
/**
 * BigInt schema based on Conway CDDL specification
 *
 * CDDL: big_int = int/ big_uint/ big_nint
 *
 * @since 2.0.0
 * @category schemas
 */
export const BigIntSchema = /*#__PURE__*/Schema.Union(
/*#__PURE__*/
// int - regular JavaScript number for small integers
Schema.TaggedStruct("Int", {
  value: Schema.Number
}),
// big_uint - positive big integers
BigUIntSchema,
// big_nint - negative big integers
BigNIntSchema);
/**
 * Creates an integer value
 *
 * @since 2.0.0
 * @category constructors
 */
export const int = value => ({
  _tag: "Int",
  value
});
/**
 * Creates a big unsigned integer from bytes
 *
 * @since 2.0.0
 * @category constructors
 */
export const bigUInt = bytes => ({
  _tag: "BigUInt",
  bytes
});
/**
 * Creates a big negative integer from bytes
 *
 * @since 2.0.0
 * @category constructors
 */
export const bigNInt = bytes => ({
  _tag: "BigNInt",
  bytes
});
/**
 * Pattern matching helper for BigInt types
 *
 * @since 2.0.0
 * @category utilities
 */
export const match = (bigInt, cases) => {
  switch (bigInt._tag) {
    case "Int":
      return cases.Int(bigInt.value);
    case "BigUInt":
      return cases.BigUInt(bigInt.bytes);
    case "BigNInt":
      return cases.BigNInt(bigInt.bytes);
    default:
      throw new Error(`Unknown BigInt variant: ${bigInt._tag}`);
  }
};
