// @evolution-sdk/evolution/dist/ByronAddress.js
import { Effect as Eff, Equal, Hash, Inspectable, ParseResult, Schema } from "effect";
import * as NetworkId from "./NetworkId.js";
/**
 * Byron legacy address format
 *
 * @since 2.0.0
 * @category schemas
 */
export class ByronAddress extends /*#__PURE__*/Schema.TaggedClass("ByronAddress")("ByronAddress", {
  networkId: NetworkId.NetworkId,
  bytes: Schema.Uint8ArrayFromSelf
}) {
  toJSON() {
    return {
      _tag: "ByronAddress",
      networkId: this.networkId,
      bytes: this.bytes
    };
  }
  toString() {
    return Inspectable.format(this.toJSON());
  }
  [Inspectable.NodeInspectSymbol]() {
    return this.toJSON();
  }
  [Equal.symbol](that) {
    return that instanceof ByronAddress && this.networkId === that.networkId && this.bytes === that.bytes;
  }
  [Hash.symbol]() {
    return Hash.hash(this.toJSON());
  }
}
/**
 * Schema for encoding/decoding Byron addresses as bytes.
 *
 * @since 2.0.0
 * @category schemas
 */
export const BytesSchema = /*#__PURE__*/Schema.transformOrFail(Schema.Uint8ArrayFromSelf, ByronAddress, {
  strict: true,
  encode: (_, __, ___, toA) => ParseResult.succeed(toA.bytes),
  decode: (_, __, ast, fromA) => Eff.gen(function* () {
    return new ByronAddress({
      networkId: NetworkId.NetworkId.make(0),
      bytes: fromA
    });
  })
});
/**
 * Schema for encoding/decoding Byron addresses as hex strings.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromHex = /*#__PURE__*/Schema.compose(Schema.Uint8ArrayFromHex, BytesSchema);
