// @evolution-sdk/evolution/dist/Bytes64.js
import { Schema } from "effect";
import * as Bytes from "./Bytes.js";
/**
 * Constant bytes length
 *
 * @since 2.0.0
 * @category constants
 */
export const BYTES_LENGTH = 64;
export const BytesFromHex = /*#__PURE__*/Schema.Uint8ArrayFromHex.pipe(/*#__PURE__*/Bytes.bytesLengthEquals(BYTES_LENGTH));
export const VariableBytesFromHex = /*#__PURE__*/Schema.Uint8ArrayFromHex.pipe(/*#__PURE__*/Bytes.bytesLengthBetween(0, BYTES_LENGTH));
// =============================================================================
// Public (throwing) API
// =============================================================================
/**
 * Decode fixed-length hex into bytes.
 *
 * @since 2.0.0
 * @category decoding
 */
export const fromHex = /*#__PURE__*/Schema.decodeSync(BytesFromHex);
/**
 * Encode fixed-length bytes to hex.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toHex = /*#__PURE__*/Schema.encodeSync(BytesFromHex);
/**
 * Decode variable-length hex (0..BYTES_LENGTH) into bytes.
 *
 * @since 2.0.0
 * @category decoding
 */
export const fromVariableHex = /*#__PURE__*/Schema.decodeSync(VariableBytesFromHex);
/**
 * Encode variable-length bytes (0..BYTES_LENGTH) to hex.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toVariableHex = /*#__PURE__*/Schema.encodeSync(VariableBytesFromHex);
export const equals = Bytes.equals;
