// @evolution-sdk/evolution/dist/Codec.js
/**
 * Simple codec factory for creating encoding utilities from schema records
 *
 * @since 2.0.0
 * @category utilities
 */
import { Effect, Either, Schema } from "effect";
/**
 * Creates encoding and decoding utilities with custom error mapping
 *
 * @since 2.0.0
 * @category constructors
 */
export const createEncoders = (schemas, ErrorClass) => {
  const createSyncMethods = (fn, operationType) => Object.fromEntries(Object.entries(schemas).map(([key, schema]) => [key, input => {
    try {
      return fn(schema)(input);
    } catch (cause) {
      throw new ErrorClass({
        message: `Failed to ${operationType} ${key}`,
        cause
      });
    }
  }]));
  const createEffectMethods = (fn, operationType) => Object.fromEntries(Object.entries(schemas).map(([key, schema]) => [key, input => fn(schema)(input).pipe(Effect.mapError(cause => new ErrorClass({
    message: `Failed to ${operationType} ${key}`,
    cause
  })))]));
  const createEitherMethods = (fn, operationType) => Object.fromEntries(Object.entries(schemas).map(([key, schema]) => [key, input => fn(schema)(input).pipe(Either.mapLeft(cause => new ErrorClass({
    message: `Failed to ${operationType} ${key}`,
    cause
  })))]));
  const encodeSync = createSyncMethods(Schema.encodeSync, "encode");
  const decodeSync = createSyncMethods(Schema.decodeSync, "decode");
  const encodeEffect = createEffectMethods(Schema.encode, "encode");
  const decodeEffect = createEffectMethods(Schema.decode, "decode");
  const encodeEither = createEitherMethods(Schema.encodeEither, "encode");
  const decodeEither = createEitherMethods(Schema.decodeEither, "decode");
  return {
    Encode: encodeSync,
    Decode: decodeSync,
    EncodeEffect: encodeEffect,
    DecodeEffect: decodeEffect,
    EncodeEither: encodeEither,
    DecodeEither: decodeEither
  };
};
