// @evolution-sdk/evolution/dist/Constitution.js
import { Effect as Eff, Equal, FastCheck, Hash, Inspectable, ParseResult, Schema } from "effect";
import * as Anchor from "./Anchor.js";
import * as CBOR from "./CBOR.js";
import * as ScriptHash from "./ScriptHash.js";
/**
 * Constitution per CDDL:
 * constitution = [anchor, script_hash/ nil]
 *
 * @since 2.0.0
 * @category schemas
 */
export class Constitution extends /*#__PURE__*/Schema.TaggedClass()("Constitution", {
  anchor: Anchor.Anchor,
  scriptHash: /*#__PURE__*/Schema.NullOr(ScriptHash.ScriptHash)
}) {
  toJSON() {
    return {
      _tag: "Constitution",
      anchor: this.anchor.toJSON(),
      scriptHash: this.scriptHash ? this.scriptHash.toJSON() : null
    };
  }
  toString() {
    return Inspectable.format(this.toJSON());
  }
  [Inspectable.NodeInspectSymbol]() {
    return this.toJSON();
  }
  [Equal.symbol](that) {
    return that instanceof Constitution && Equal.equals(this.anchor, that.anchor) && Equal.equals(this.scriptHash, that.scriptHash);
  }
  [Hash.symbol]() {
    return Hash.cached(this, Hash.combine(Hash.hash("Constitution"))(Hash.combine(Hash.hash(this.anchor))(Hash.hash(this.scriptHash))));
  }
}
/**
 * CDDL tuple schema for Constitution
 */
export const CDDLSchema = /*#__PURE__*/Schema.Tuple(Anchor.CDDLSchema,
/*#__PURE__*/
// anchor
Schema.NullOr(CBOR.ByteArray) // script_hash / nil
);
/**
 * Transform between CDDL tuple and typed Constitution
 */
export const FromCDDL = /*#__PURE__*/Schema.transformOrFail(CDDLSchema, /*#__PURE__*/Schema.typeSchema(Constitution), {
  strict: true,
  encode: toA => Eff.gen(function* () {
    const anchor = yield* ParseResult.encode(Anchor.FromCDDL)(toA.anchor);
    const scriptHash = toA.scriptHash ? yield* ParseResult.encode(ScriptHash.FromBytes)(toA.scriptHash) : null;
    return [anchor, scriptHash];
  }),
  decode: ([anchorTuple, scriptHashBytes]) => Eff.gen(function* () {
    const anchor = yield* ParseResult.decode(Anchor.FromCDDL)(anchorTuple);
    const scriptHash = scriptHashBytes ? yield* ParseResult.decode(ScriptHash.FromBytes)(scriptHashBytes) : null;
    return new Constitution({
      anchor,
      scriptHash
    });
  })
});
export const FromCBORBytes = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.compose(CBOR.FromBytes(options), FromCDDL);
export const FromCBORHex = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.compose(CBOR.FromHex(options), FromCBORBytes(options));
/**
 * Arbitrary for Constitution
 */
export const arbitrary = /*#__PURE__*/FastCheck.tuple(Anchor.arbitrary, FastCheck.option(ScriptHash.arbitrary, {
  nil: null
})).map(([anchor, scriptHash]) => new Constitution({
  anchor,
  scriptHash
}, {
  disableValidation: true
}));
// ============================================================================
// Decoding Functions
// ============================================================================
/**
 * Parse Constitution from CBOR bytes.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromCBORBytes = (bytes, options) => Schema.decodeSync(FromCBORBytes(options))(bytes);
/**
 * Parse Constitution from CBOR hex string.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromCBORHex = (hex, options) => Schema.decodeSync(FromCBORHex(options))(hex);
// ============================================================================
// Encoding Functions
// ============================================================================
/**
 * Convert Constitution to CBOR bytes.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toCBORBytes = (constitution, options) => Schema.encodeSync(FromCBORBytes(options))(constitution);
/**
 * Convert Constitution to CBOR hex string.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toCBORHex = (constitution, options) => Schema.encodeSync(FromCBORHex(options))(constitution);
