// @evolution-sdk/evolution/dist/DataJson.js
import { pipe, Schema } from "effect";
// Schema for Plutus data types
const renderParseIssue = issue => typeof issue.actual === "object" ? "[complex value]" : String(issue.actual);
const HEX_REGEX = /^(?:[0-9a-f]{2})*$/;
const HexString = self => pipe(self, Schema.filter(value => HEX_REGEX.test(value), {
  message: issue => `Expected a hexadecimal string but received: ${issue.actual}.`
}));
export const ByteArray = /*#__PURE__*/Schema.Struct({
  bytes: pipe(Schema.String, Schema.annotations({
    message: issue => ({
      message: `Expected ByteArray but got ${renderParseIssue(issue)}.`,
      override: true
    })
  }), HexString)
}).annotations({
  identifier: "ByteArray"
});
export const Integer = /*#__PURE__*/Schema.Struct({
  int: Schema.Number
}).annotations({
  identifier: "Integer",
  message: issue => {
    return `Expected Integer but got ${renderParseIssue(issue)}.`;
  }
});
export const isInteger = /*#__PURE__*/Schema.is(Integer);
export const isByteArray = /*#__PURE__*/Schema.is(ByteArray);
export const List = /*#__PURE__*/Schema.Array(Schema.suspend(() => Data)).annotations({
  identifier: "List",
  message: issue => {
    return `Expected List but got ${renderParseIssue(issue)}.`;
  }
});
export const isList = /*#__PURE__*/Schema.is(List);
export const Map = /*#__PURE__*/Schema.Record({
  key: Schema.String,
  value: Schema.suspend(() => Data)
}).annotations({
  identifier: "Map",
  message: issue => {
    return `Expected Map but got ${renderParseIssue(issue)}.`;
  }
});
export const isMap = /*#__PURE__*/Schema.is(Map);
export const Constr = /*#__PURE__*/Schema.Struct({
  constructor: Schema.Number,
  fields: Schema.Array(Schema.suspend(() => Data))
}).annotations({
  identifier: "Constr",
  message: issue => {
    return `Expected Constr but got ${renderParseIssue(issue)}.`;
  }
});
export const isConstr = /*#__PURE__*/Schema.is(Constr);
// export const Data = Schema.Union(Integer, ByteArray, List, Map, Constr);
export const Data = /*#__PURE__*/Schema.Union(Integer, ByteArray, List, Map, Constr);
