// @evolution-sdk/evolution/dist/DnsName.js
import { Schema } from "effect";
import * as Text128 from "./Text128.js";
/**
 * Schema for DnsName with DNS-specific validation.
 * dns_name = text .size (0 .. 128)
 *
 * @since 2.0.0
 * @category model
 */
export const DnsName = /*#__PURE__*/Text128.Text128.pipe(/*#__PURE__*/Schema.brand("DnsName"));
export const FromBytes = /*#__PURE__*/Schema.compose(Text128.FromBytes, DnsName).annotations({
  identifier: "DnsName.FromBytes"
});
export const FromHex = /*#__PURE__*/Schema.compose(Text128.FromHex, DnsName).annotations({
  identifier: "DnsName.FromHex"
});
/**
 * Check if the given value is a valid DnsName
 *
 * @since 2.0.0
 * @category predicates
 */
export const isDnsName = /*#__PURE__*/Schema.is(DnsName);
/**
 * FastCheck arbitrary for generating random DnsName instances.
 *
 * @since 2.0.0
 * @category arbitrary
 */
export const arbitrary = /*#__PURE__*/Text128.arbitrary.map(text => DnsName.make(text));
// ============================================================================
// Root Functions
// ============================================================================
/**
 * Parse DnsName from bytes.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromBytes = /*#__PURE__*/Schema.decodeSync(FromBytes);
/**
 * Parse DnsName from hex string.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromHex = /*#__PURE__*/Schema.decodeSync(FromHex);
/**
 * Encode DnsName to bytes.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toBytes = /*#__PURE__*/Schema.encodeSync(FromBytes);
/**
 * Encode DnsName to hex string.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toHex = /*#__PURE__*/Schema.encodeSync(FromHex);
