// @evolution-sdk/evolution/dist/Ed25519Signature.js
import { Equal, FastCheck, Hash, Inspectable, Schema } from "effect";
import * as Bytes from "./Bytes.js";
import * as Bytes64 from "./Bytes64.js";
/**
 * Class-based Ed25519Signature with compile-time and runtime safety.
 * ed25519_signature = bytes .size 64
 * Follows the Conway-era CDDL specification.
 *
 * @since 2.0.0
 * @category model
 */
export class Ed25519Signature extends /*#__PURE__*/Schema.Class("Ed25519Signature")({
  bytes: Bytes64.BytesFromHex
}) {
  toJSON() {
    return {
      _tag: "Ed25519Signature",
      bytes: Bytes.toHex(this.bytes)
    };
  }
  toString() {
    return Inspectable.format(this.toJSON());
  }
  [Inspectable.NodeInspectSymbol]() {
    return this.toJSON();
  }
  [Equal.symbol](that) {
    return that instanceof Ed25519Signature && Bytes.equals(this.bytes, that.bytes);
  }
  [Hash.symbol]() {
    return Hash.array(Array.from(this.bytes));
  }
}
/**
 * Schema transformer from bytes to Ed25519Signature.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromBytes = /*#__PURE__*/Schema.transform(Schema.typeSchema(Bytes64.BytesFromHex), Schema.typeSchema(Ed25519Signature), {
  strict: true,
  decode: bytes => new Ed25519Signature({
    bytes
  }, {
    disableValidation: true
  } // Disable validation since we already check length in Bytes64
  ),
  encode: signature => new Uint8Array(signature.bytes)
}).annotations({
  identifier: "Ed25519Signature.FromBytes"
});
/**
 * Schema transformer from hex string to Ed25519Signature.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromHex = /*#__PURE__*/Schema.compose(Bytes64.BytesFromHex,
// string -> Bytes64
FromBytes).annotations({
  identifier: "Ed25519Signature.FromHex"
});
// ============================================================================
// Core Functions (functional interface)
// ============================================================================
/**
 * Parse Ed25519Signature from bytes (unsafe - throws on error).
 *
 * @since 2.0.0
 * @category constructors
 */
export const fromBytes = /*#__PURE__*/Schema.decodeSync(FromBytes);
/**
 * Parse Ed25519Signature from hex string (unsafe - throws on error).
 *
 * @since 2.0.0
 * @category constructors
 */
export const fromHex = /*#__PURE__*/Schema.decodeSync(FromHex);
/**
 * Convert to hex string using optimized lookup table.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toHex = /*#__PURE__*/Schema.encodeSync(FromHex);
/**
 * Get the underlying bytes (returns a copy for safety).
 *
 * @since 2.0.0
 * @category encoding
 */
export const toBytes = /*#__PURE__*/Schema.encodeSync(FromBytes);
/**
 * Check if value is an Ed25519Signature instance.
 *
 * @since 2.0.0
 * @category predicates
 */
export const is = /*#__PURE__*/Schema.is(Ed25519Signature);
/**
 * FastCheck arbitrary for generating random Ed25519Signature instances.
 *
 * @since 2.0.0
 * @category arbitrary
 */
export const arbitrary = /*#__PURE__*/FastCheck.uint8Array({
  minLength: 64,
  maxLength: 64
}).map(bytes => new Ed25519Signature({
  bytes
}, {
  disableValidation: true
}));
