// @evolution-sdk/evolution/dist/Function.js
import { Either, Schema } from "effect";
import * as Bytes from "./Bytes.js";
import * as CBOR from "./CBOR.js";
/**
 * Creates a named function with proper stack traces using dynamic object property
 */
export const makeDecodeSync = (schemaTransformer, ErrorClass, functionName) => {
  // Pre-create decoder once to avoid per-call allocation/closure overhead
  const decode = Schema.decodeEither(schemaTransformer);
  // Create a named function using object property syntax
  const fn = {
    [functionName](input) {
      const result = decode(input);
      if (result._tag === "Left") {
        const error = new ErrorClass({
          message: `Failed to decode in ${functionName}`,
          cause: result.left
        });
        if (Error.captureStackTrace) {
          Error.captureStackTrace(error, fn[functionName]);
        }
        throw error;
      }
      return result.right;
    }
  };
  return fn[functionName];
};
export const makeEncodeSync = (schemaTransformer, ErrorClass, functionName) => {
  // Pre-create encoder once to avoid per-call allocation/closure overhead
  const encode = Schema.encodeEither(schemaTransformer);
  const fn = {
    [functionName]: input => {
      const result = encode(input);
      if (result._tag === "Left") {
        const error = new ErrorClass({
          message: `Failed to encode in ${functionName}`,
          cause: result.left
        });
        if (Error.captureStackTrace) {
          Error.captureStackTrace(error, fn[functionName]);
        }
        throw error;
      }
      return result.right;
    }
  };
  return fn[functionName];
};
/**
 * Creates a synchronous function that encodes a value to CBOR bytes.
 * Combines schema encoding with CBOR serialization in one step.
 *
 * @since 2.0.0
 * @category constructors
 */
export const makeCBOREncodeSync = (schemaTransformer, ErrorClass, functionName, defaultOptions) => {
  // Schema.encodeSync goes from T -> A (opposite direction of the schema)
  const encode = Schema.encodeSync(schemaTransformer);
  const fn = {
    [functionName]: (input, options) => {
      try {
        const cborValue = encode(input);
        return CBOR.internalEncodeSync(cborValue, options || defaultOptions);
      } catch (e) {
        const error = new ErrorClass({
          message: `Failed to encode in ${functionName}`,
          cause: e
        });
        if (Error.captureStackTrace) {
          Error.captureStackTrace(error, wrapped);
        }
        throw error;
      }
    }
  };
  const wrapped = fn[functionName];
  return wrapped;
};
export const makeDecodeEither = (schema, ErrorClass) => input => Schema.decodeEither(schema)(input).pipe(Either.mapLeft(e => new ErrorClass({
  message: e.message,
  cause: e
})));
export const makeEncodeEither = (schema, ErrorClass) => input => Schema.encodeEither(schema)(input).pipe(Either.mapLeft(e => new ErrorClass({
  message: e.message,
  cause: e
})));
/**
 * Creates a synchronous function that decodes CBOR bytes into a value using a schema.
 * This pairs with makeCBOREncodeSync and avoids extra FromCBOR* transformers by using FromCDDL directly.
 *
 * @since 2.0.0
 * @category constructors
 */
export const makeCBORDecodeSync = (schemaTransformer, ErrorClass, functionName, defaultOptions) => {
  // Pre-create decoder once (T -> A)
  const decode = Schema.decodeSync(schemaTransformer);
  const fn = {
    [functionName]: (bytes, options) => {
      try {
        const cborValue = CBOR.internalDecodeSync(bytes, options || defaultOptions);
        return decode(cborValue);
      } catch (e) {
        const error = new ErrorClass({
          message: `Failed to decode in ${functionName}`,
          cause: e
        });
        if (Error.captureStackTrace) {
          Error.captureStackTrace(error, wrapped);
        }
        throw error;
      }
    }
  };
  const wrapped = fn[functionName];
  return wrapped;
};
/**
 * Creates a synchronous function that encodes a value to CBOR hex string.
 * Uses a schema to encode T -> A then serializes to hex.
 */
export const makeCBOREncodeHexSync = (schemaTransformer, ErrorClass, functionName, defaultOptions) => {
  const encode = Schema.encodeSync(schemaTransformer);
  const fn = {
    [functionName]: (input, options) => {
      try {
        const cborValue = encode(input);
        const bytes = CBOR.internalEncodeSync(cborValue, options || defaultOptions);
        return Bytes.toHex(bytes);
      } catch (e) {
        const error = new ErrorClass({
          message: `Failed to encode hex in ${functionName}`,
          cause: e
        });
        if (Error.captureStackTrace) {
          Error.captureStackTrace(error, wrapped);
        }
        throw error;
      }
    }
  };
  const wrapped = fn[functionName];
  return wrapped;
};
/**
 * Creates a synchronous function that decodes a CBOR hex string into a value using a schema.
 */
export const makeCBORDecodeHexSync = (schemaTransformer, ErrorClass, functionName, defaultOptions) => {
  const decode = Schema.decodeSync(schemaTransformer);
  const fn = {
    [functionName]: (hex, options) => {
      try {
        const bytes = Bytes.fromHex(hex);
        const cborValue = CBOR.internalDecodeSync(bytes, options || defaultOptions);
        return decode(cborValue);
      } catch (e) {
        const error = new ErrorClass({
          message: `Failed to decode hex in ${functionName}`,
          cause: e
        });
        if (Error.captureStackTrace) {
          Error.captureStackTrace(error, wrapped);
        }
        throw error;
      }
    }
  };
  const wrapped = fn[functionName];
  return wrapped;
};
/**
 * Creates a function that decodes CBOR bytes into a value using a schema, returning Either.
 */
export const makeCBORDecodeEither = (schemaTransformer, ErrorClass, defaultOptions) => (bytes, options) => Either.try(() => CBOR.internalDecodeSync(bytes, options || defaultOptions)).pipe(Either.flatMap(cbor => Schema.decodeEither(schemaTransformer)(cbor)), Either.mapLeft(e => new ErrorClass({
  message: e.message,
  cause: e
})));
/**
 * Creates a function that decodes CBOR hex string into a value using a schema, returning Either.
 */
export const makeCBORDecodeHexEither = (schemaTransformer, ErrorClass, defaultOptions) => (hex, options) => Either.try(() => Bytes.fromHex(hex)).pipe(Either.flatMap(bytes => Either.try(() => CBOR.internalDecodeSync(bytes, options || defaultOptions))), Either.flatMap(cbor => Schema.decodeEither(schemaTransformer)(cbor)), Either.mapLeft(e => new ErrorClass({
  message: e.message,
  cause: e
})));
/**
 * Creates a function that encodes a value to CBOR bytes using a schema, returning Either.
 */
export const makeCBOREncodeEither = (schemaTransformer, ErrorClass, defaultOptions) => (value, options) => Schema.encodeEither(schemaTransformer)(value).pipe(Either.flatMap(cbor => Either.try(() => CBOR.internalEncodeSync(cbor, options || defaultOptions))), Either.mapLeft(e => new ErrorClass({
  message: e.message,
  cause: e
})));
/**
 * Creates a function that encodes a value to CBOR hex string using a schema, returning Either.
 */
export const makeCBOREncodeHexEither = (schemaTransformer, ErrorClass, defaultOptions) => (input, options) => Schema.encodeEither(schemaTransformer)(input).pipe(Either.flatMap(cborValue => Either.try(() => CBOR.internalEncodeSync(cborValue, options || defaultOptions))), Either.map(bytes => Bytes.toHex(bytes)), Either.mapLeft(e => new ErrorClass({
  message: e.message,
  cause: e
})));
