// @evolution-sdk/evolution/dist/KesSignature.js
import { Equal, FastCheck, Hash, Inspectable, Schema } from "effect";
import * as Bytes from "./Bytes.js";
import * as Bytes448 from "./Bytes448.js";
/**
 * KesSignature model stored as 448 raw bytes.
 * kes_signature = bytes .size 448
 *
 * @since 2.0.0
 * @category schemas
 */
export class KesSignature extends /*#__PURE__*/Schema.TaggedClass()("KesSignature", {
  bytes: Bytes448.BytesFromHex
}) {
  toJSON() {
    return {
      _tag: "KesSignature",
      bytes: Bytes.toHex(this.bytes)
    };
  }
  toString() {
    return Inspectable.format(this.toJSON());
  }
  [Inspectable.NodeInspectSymbol]() {
    return this.toJSON();
  }
  [Equal.symbol](that) {
    return that instanceof KesSignature && Bytes.equals(this.bytes, that.bytes);
  }
  [Hash.symbol]() {
    return Hash.array(Array.from(this.bytes));
  }
}
// Transform between raw bytes (Uint8Array length 448) and KesSignature
export const FromBytes = /*#__PURE__*/Schema.transform(Schema.typeSchema(Bytes448.BytesFromHex), Schema.typeSchema(KesSignature), {
  strict: true,
  decode: bytes => new KesSignature({
    bytes
  }, {
    disableValidation: true
  }),
  encode: value => value.bytes
}).annotations({
  identifier: "KesSignature.FromBytes"
});
export const FromHex = /*#__PURE__*/Schema.compose(Bytes448.BytesFromHex,
// string -> Uint8Array(448)
FromBytes // bytes -> KesSignature
).annotations({
  identifier: "KesSignature.FromHex"
});
/**
 * Predicate for KesSignature instances
 *
 * @since 2.0.0
 * @category predicates
 */
export const isKesSignature = /*#__PURE__*/Schema.is(KesSignature);
/**
 * FastCheck arbitrary for generating random KesSignature instances.
 *
 * @since 2.0.0
 * @category arbitrary
 */
export const arbitrary = /*#__PURE__*/FastCheck.uint8Array({
  minLength: 448,
  maxLength: 448
}).map(bytes => new KesSignature({
  bytes
}, {
  disableValidation: true
}));
// ============================================================================
// Root Functions
// ============================================================================
/**
 * Parse KesSignature from bytes.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromBytes = /*#__PURE__*/Schema.decodeSync(FromBytes);
/**
 * Parse KesSignature from hex string.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromHex = /*#__PURE__*/Schema.decodeSync(FromHex);
/**
 * Encode KesSignature to bytes.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toBytes = /*#__PURE__*/Schema.encodeSync(FromBytes);
/**
 * Encode KesSignature to hex string.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toHex = /*#__PURE__*/Schema.encodeSync(FromHex);
