// @evolution-sdk/evolution/dist/Language.js
import { FastCheck, Schema } from "effect";
/**
 * Plutus languages supported in cost models.
 *
 * CDDL: language = 0 / 1 / 2  ; plutus_v1 / plutus_v2 / plutus_v3
 *
 * @since 2.0.0
 */
export const Language = /*#__PURE__*/Schema.Literal("PlutusV1", "PlutusV2", "PlutusV3");
export const CDDLSchema = /*#__PURE__*/Schema.Literal(0n, 1n, 2n);
export const FromCDDL = /*#__PURE__*/Schema.transform(CDDLSchema, /*#__PURE__*/Schema.typeSchema(Language), {
  strict: true,
  encode: lang => {
    switch (lang) {
      case "PlutusV1":
        return 0n;
      case "PlutusV2":
        return 1n;
      case "PlutusV3":
        return 2n;
    }
  },
  decode: tag => {
    switch (tag) {
      case 0n:
        return "PlutusV1";
      case 1n:
        return "PlutusV2";
      case 2n:
        return "PlutusV3";
    }
  }
});
export const arbitrary = /*#__PURE__*/FastCheck.oneof(/*#__PURE__*/FastCheck.constant("PlutusV1"), /*#__PURE__*/FastCheck.constant("PlutusV2"), /*#__PURE__*/FastCheck.constant("PlutusV3"));
// Simple helpers
export const toCDDL = lang => Schema.encodeSync(FromCDDL)(lang);
export const fromCDDL = tag => Schema.decodeSync(FromCDDL)(tag);
