// @evolution-sdk/evolution/dist/Metadata.js
import { FastCheck, ParseResult, Schema } from "effect";
import * as CBOR from "./CBOR.js";
import * as Numeric from "./Numeric.js";
import * as TransactionMetadatum from "./TransactionMetadatum.js";
/**
 * Schema for transaction metadatum label (uint64 per Cardano CDDL spec).
 * Labels must be in range 0 to 2^64-1.
 *
 * @since 2.0.0
 * @category schemas
 */
export const MetadataLabel = /*#__PURE__*/Numeric.Uint64Schema.annotations({
  identifier: "Metadata.MetadataLabel",
  description: "A transaction metadatum label (0 to 2^64-1)"
});
/**
 * Schema for transaction metadata (map from labels to metadata).
 * ```
 * Represents: metadata = {* transaction_metadatum_label => transaction_metadatum}
 * ```
 *
 * @since 2.0.0
 * @category schemas
 */
export const Metadata = /*#__PURE__*/Schema.Map({
  key: MetadataLabel,
  value: TransactionMetadatum.TransactionMetadatumSchema
}).annotations({
  identifier: "Metadata",
  description: "Transaction metadata as a map from labels to transaction metadata values"
});
/**
 * CDDL schema for Metadata (CBOR-compatible representation).
 * Maps bigint labels to transaction metadatum values.
 *
 * Uses Schema.typeSchema(TransactionMetadatumSchema) because CBOR decoding
 * returns runtime types (bigint, Uint8Array, Map) directly.
 *
 * @since 2.0.0
 * @category schemas
 */
export const CDDLSchema = /*#__PURE__*/Schema.MapFromSelf({
  key: CBOR.Integer,
  // MetadataLabel as bigint
  value: /*#__PURE__*/Schema.suspend(() => Schema.typeSchema(TransactionMetadatum.TransactionMetadatumSchema))
});
/**
 * Transform schema from CDDL to Metadata.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromCDDL = /*#__PURE__*/Schema.transformOrFail(CDDLSchema, Schema.typeSchema(Metadata), {
  strict: true,
  encode: metadata => ParseResult.succeed(new Map(metadata)),
  decode: cddl => ParseResult.succeed(new Map(cddl))
}).annotations({
  identifier: "Metadata.FromCDDL",
  description: "Transforms CBOR structure to Metadata"
});
/**
 * Schema transformer for Metadata from CBOR bytes.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromCBORBytes = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.compose(CBOR.FromBytes(options),
// Uint8Array → CBOR
FromCDDL // CBOR → Metadata
).annotations({
  identifier: "Metadata.FromCBORBytes",
  description: "Transforms CBOR bytes to Metadata"
});
/**
 * Schema transformer for Metadata from CBOR hex string.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromCBORHex = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.compose(Schema.Uint8ArrayFromHex,
// string → Uint8Array
FromCBORBytes(options) // Uint8Array → Metadata
).annotations({
  identifier: "Metadata.FromCBORHex",
  description: "Transforms CBOR hex string to Metadata"
});
/**
 * FastCheck arbitrary for generating random Metadata instances.
 *
 * @since 2.0.0
 * @category testing
 */
export const arbitrary = /*#__PURE__*/FastCheck.array(FastCheck.tuple(FastCheck.bigInt({
  min: 0n,
  max: 255n
}),
// MetadataLabel (uint8)
TransactionMetadatum.arbitrary), {
  maxLength: 5
}).map(entries => fromEntries(entries));
// ============================================================================
// Parsing Functions
// ============================================================================
/**
 * Parse Metadata from CBOR bytes.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromCBORBytes = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.decodeSync(Schema.compose(CBOR.FromBytes(options), FromCDDL));
/**
 * Parse Metadata from CBOR hex string.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromCBORHex = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.decodeSync(Schema.compose(CBOR.FromHex(options), FromCBORBytes(options)));
// ============================================================================
// Encoding Functions
// ============================================================================
/**
 * Convert Metadata to CBOR bytes.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toCBORBytes = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.encodeSync(FromCBORBytes(options));
/**
 * Convert Metadata to CBOR hex string.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toCBORHex = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.encodeSync(FromCBORHex(options));
// ============================================================================
// Factory Functions
// ============================================================================
/**
 * Create Metadata from an array of label-metadatum pairs.
 *
 * @since 2.0.0
 * @category constructors
 */
export const fromEntries = entries => new Map(entries);
/**
 * Create an empty Metadata map.
 *
 * @since 2.0.0
 * @category constructors
 */
export const empty = () => new Map();
/**
 * Add or update a metadata entry.
 *
 * @since 2.0.0
 * @category constructors
 */
export const set = (metadata, label, metadatum) => {
  const newMap = new Map(metadata);
  newMap.set(label, metadatum);
  return newMap;
};
/**
 * Get a metadata entry by label.
 *
 * @since 2.0.0
 * @category utilities
 */
export const get = (metadata, label) => metadata.get(label);
/**
 * Check if a label exists in the metadata.
 *
 * @since 2.0.0
 * @category utilities
 */
export const has = (metadata, label) => metadata.has(label);
/**
 * Remove a metadata entry by label.
 *
 * @since 2.0.0
 * @category constructors
 */
export const remove = (metadata, label) => {
  const newMap = new Map(metadata);
  newMap.delete(label);
  return newMap;
};
/**
 * Get the size (number of entries) of the metadata.
 *
 * @since 2.0.0
 * @category utilities
 */
export const size = metadata => metadata.size;
/**
 * Get all labels in the metadata.
 *
 * @since 2.0.0
 * @category utilities
 */
export const labels = metadata => Array.from(metadata.keys());
/**
 * Get all metadata values in the metadata.
 *
 * @since 2.0.0
 * @category utilities
 */
export const values = metadata => Array.from(metadata.values());
/**
 * Get all entries in the metadata.
 *
 * @since 2.0.0
 * @category utilities
 */
export const entries = metadata => Array.from(metadata.entries());
