// @evolution-sdk/evolution/dist/NativeScriptsOLD.js
import { Data, Either as E, FastCheck, ParseResult, Schema } from "effect";
import * as CBOR from "./CBOR.js";
import * as Function from "./Function.js";
/**
 * Error class for Native script related operations.
 *
 * @since 2.0.0
 * @category errors
 */
export class NativeError extends /*#__PURE__*/Data.TaggedError("NativeError") {}
/**
 * Represents a cardano-cli JSON script syntax
 *
 * Native type follows the standard described in the
 * link https://github.com/IntersectMBO/cardano-node/blob/1.26.1-with-cardano-cli/doc/reference/simple-scripts.md#json-script-syntax JSON script syntax documentation.
 *
 * @since 2.0.0
 * @category schemas
 */
export const NativeSchema = /*#__PURE__*/Schema.Union(Schema.Struct({
  type: Schema.Literal("sig"),
  keyHash: Schema.String
}), Schema.Struct({
  type: Schema.Literal("before"),
  slot: Schema.Number
}), Schema.Struct({
  type: Schema.Literal("after"),
  slot: Schema.Number
}), Schema.Struct({
  type: Schema.Literal("all"),
  scripts: Schema.Array(Schema.suspend(() => NativeSchema))
}), Schema.Struct({
  type: Schema.Literal("any"),
  scripts: Schema.Array(Schema.suspend(() => NativeSchema))
}), Schema.Struct({
  type: Schema.Literal("atLeast"),
  required: Schema.Number,
  scripts: Schema.Array(Schema.suspend(() => NativeSchema))
})).annotations({
  identifier: "Native",
  title: "Native Script",
  description: "A native script following cardano-cli JSON syntax"
});
export const Native = NativeSchema;
/**
 * Smart constructor for Native that validates and applies branding.
 *
 * @since 2.0.0
 * @category constructors
 */
export const make = native => native;
/**
 * CDDL schemas for native scripts.
 *
 * These schemas define the CBOR encoding format for native scripts according to the CDDL specification:
 *
 * - script_pubkey = (0, addr_keyhash)
 * - script_all = (1, [* native_script])
 * - script_any = (2, [* native_script])
 * - script_n_of_k = (3, n : int64, [* native_script])
 * - invalid_before = (4, slot_no)
 * - invalid_hereafter = (5, slot_no)
 * - slot_no = uint .size 8
 *
 * @since 2.0.0
 * @category schemas
 */
const ScriptPubKeyCDDL = /*#__PURE__*/Schema.Tuple(/*#__PURE__*/Schema.Literal(0n), Schema.Uint8ArrayFromSelf);
const ScriptAllCDDL = /*#__PURE__*/Schema.Tuple(/*#__PURE__*/Schema.Literal(1n), /*#__PURE__*/Schema.Array(/*#__PURE__*/Schema.suspend(() => Schema.encodedSchema(FromCDDL))));
const ScriptAnyCDDL = /*#__PURE__*/Schema.Tuple(/*#__PURE__*/Schema.Literal(2n), /*#__PURE__*/Schema.Array(/*#__PURE__*/Schema.suspend(() => Schema.encodedSchema(FromCDDL))));
const ScriptNOfKCDDL = /*#__PURE__*/Schema.Tuple(/*#__PURE__*/Schema.Literal(3n), CBOR.Integer, /*#__PURE__*/Schema.Array(/*#__PURE__*/Schema.suspend(() => Schema.encodedSchema(FromCDDL))));
const InvalidBeforeCDDL = /*#__PURE__*/Schema.Tuple(/*#__PURE__*/Schema.Literal(4n), CBOR.Integer);
const InvalidHereafterCDDL = /*#__PURE__*/Schema.Tuple(/*#__PURE__*/Schema.Literal(5n), CBOR.Integer);
export const CDDLSchema = /*#__PURE__*/Schema.Union(ScriptPubKeyCDDL, ScriptAllCDDL, ScriptAnyCDDL, ScriptNOfKCDDL, InvalidBeforeCDDL, InvalidHereafterCDDL);
/**
 * Schema for NativeCDDL union type.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromCDDL = /*#__PURE__*/Schema.transformOrFail(CDDLSchema, /*#__PURE__*/Schema.typeSchema(Native), {
  strict: true,
  encode: native => internalEncodeCDDL(native),
  decode: cborTuple => internalDecodeCDDL(cborTuple)
});
/**
 * Convert a Native to its CDDL representation.
 *
 * @since 2.0.0
 * @category encoding
 */
export const internalEncodeCDDL = native => E.gen(function* () {
  switch (native.type) {
    case "sig":
      {
        // Convert hex string keyHash to bytes for CBOR encoding
        const keyHashBytes = yield* ParseResult.decodeEither(Schema.Uint8ArrayFromHex)(native.keyHash);
        return [0n, keyHashBytes];
      }
    case "all":
      {
        const scriptResults = [];
        for (const script of native.scripts) {
          const encoded = yield* internalEncodeCDDL(script);
          scriptResults.push(encoded);
        }
        return [1n, scriptResults];
      }
    case "any":
      {
        const scriptResults = [];
        for (const script of native.scripts) {
          const encoded = yield* internalEncodeCDDL(script);
          scriptResults.push(encoded);
        }
        return [2n, scriptResults];
      }
    case "atLeast":
      {
        const scriptResults = [];
        for (const script of native.scripts) {
          const encoded = yield* internalEncodeCDDL(script);
          scriptResults.push(encoded);
        }
        return [3n, BigInt(native.required), scriptResults];
      }
    case "before":
      {
        return [4n, BigInt(native.slot)];
      }
    case "after":
      {
        return [5n, BigInt(native.slot)];
      }
  }
});
/**
 * Convert a CDDL representation back to a Native.
 *
 * This function recursively decodes nested CBOR scripts and constructs
 * the appropriate Native instances.
 *
 * @since 2.0.0
 * @category decoding
 */
export const internalDecodeCDDL = cborTuple => E.gen(function* () {
  switch (cborTuple[0]) {
    case 0n:
      {
        // sig: [0, keyHash_bytes] - convert bytes back to hex string
        const [, keyHashBytes] = cborTuple;
        const keyHash = yield* ParseResult.encodeEither(Schema.Uint8ArrayFromHex)(keyHashBytes);
        return {
          type: "sig",
          keyHash
        };
      }
    case 1n:
      {
        // all: [1, [native_script, ...]]
        const [, scriptCBORs] = cborTuple;
        const scripts = [];
        for (const scriptCBOR of scriptCBORs) {
          const script = yield* internalDecodeCDDL(scriptCBOR);
          scripts.push(script);
        }
        return {
          type: "all",
          scripts
        };
      }
    case 2n:
      {
        // any: [2, [native_script, ...]]
        const [, scriptCBORs] = cborTuple;
        const scripts = [];
        for (const scriptCBOR of scriptCBORs) {
          const script = yield* internalDecodeCDDL(scriptCBOR);
          scripts.push(script);
        }
        return {
          type: "any",
          scripts
        };
      }
    case 3n:
      {
        // atLeast: [3, required, [native_script, ...]]
        const [, required, scriptCBORs] = cborTuple;
        const scripts = [];
        for (const scriptCBOR of scriptCBORs) {
          const script = yield* internalDecodeCDDL(scriptCBOR);
          scripts.push(script);
        }
        return {
          type: "atLeast",
          required: Number(required),
          scripts
        };
      }
    case 4n:
      {
        // before: [4, slot]
        const [, slot] = cborTuple;
        return {
          type: "before",
          slot: Number(slot)
        };
      }
    case 5n:
      {
        // after: [5, slot]
        const [, slot] = cborTuple;
        return {
          type: "after",
          slot: Number(slot)
        };
      }
    default:
      // This should never happen with proper CBOR validation
      return yield* E.left(new ParseResult.Type(Schema.Literal(0, 1, 2, 3, 4, 5).ast, cborTuple[0]));
  }
});
/**
 * FastCheck arbitrary for Native scripts.
 * Generates valid native scripts with bounded depth and sizes.
 *
 * Depth limit prevents exponential blow-up. At depth 0, only base cases are generated.
 */
const nativeArbitrary = depth => {
  const baseSig = FastCheck.record({
    type: FastCheck.constant("sig"),
    // 28-byte keyhash (56 hex chars)
    keyHash: FastCheck.hexaString({
      minLength: 56,
      maxLength: 56
    })
  });
  const baseBefore = FastCheck.record({
    type: FastCheck.constant("before"),
    slot: FastCheck.integer({
      min: 0,
      max: 10_000_000
    })
  });
  const baseAfter = FastCheck.record({
    type: FastCheck.constant("after"),
    slot: FastCheck.integer({
      min: 0,
      max: 10_000_000
    })
  });
  if (depth <= 0) {
    return FastCheck.oneof(baseSig, baseBefore, baseAfter);
  }
  const sub = nativeArbitrary(depth - 1);
  const scriptsArray = FastCheck.array(sub, {
    minLength: 0,
    maxLength: 3
  });
  const all = scriptsArray.map(scripts => ({
    type: "all",
    scripts
  }));
  const any = scriptsArray.map(scripts => ({
    type: "any",
    scripts
  }));
  const atLeast = FastCheck.array(sub, {
    minLength: 0,
    maxLength: 4
  }).chain(scripts => FastCheck.integer({
    min: 0,
    max: scripts.length
  }).map(required => ({
    type: "atLeast",
    required,
    scripts
  })));
  // Weight base cases a bit higher for performance and balance
  return FastCheck.oneof({
    arbitrary: baseSig,
    weight: 3
  }, {
    arbitrary: baseBefore,
    weight: 2
  }, {
    arbitrary: baseAfter,
    weight: 2
  }, {
    arbitrary: all,
    weight: 1
  }, {
    arbitrary: any,
    weight: 1
  }, {
    arbitrary: atLeast,
    weight: 1
  });
};
export const arbitrary = /*#__PURE__*/nativeArbitrary(2);
/**
 * Deep structural equality for Native scripts.
 * Compares shape, values and recurses into nested scripts.
 */
export const equals = (a, b) => {
  if (a.type !== b.type) return false;
  switch (a.type) {
    case "sig":
      return a.keyHash === b.keyHash;
    case "before":
      return a.slot === b.slot;
    case "after":
      return a.slot === b.slot;
    case "all":
    case "any":
      {
        const as = a.scripts;
        const bs = b.scripts;
        if (as.length !== bs.length) return false;
        for (let i = 0; i < as.length; i++) if (!equals(as[i], bs[i])) return false;
        return true;
      }
    case "atLeast":
      {
        const bs = b;
        if (a.required !== bs.required) return false;
        const as = a.scripts;
        const bscripts = bs.scripts;
        if (as.length !== bscripts.length) return false;
        for (let i = 0; i < as.length; i++) if (!equals(as[i], bscripts[i])) return false;
        return true;
      }
  }
};
/**
 * CBOR bytes transformation schema for Native.
 * Transforms between CBOR bytes and Native using CBOR encoding.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromCBORBytes = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.compose(CBOR.FromBytes(options),
// Uint8Array → CBOR
FromCDDL // CBOR → Native
).annotations({
  identifier: "Native.FromCBORBytes",
  title: "Native from CBOR Bytes",
  description: "Transforms CBOR bytes to Native"
});
/**
 * CBOR hex transformation schema for Native.
 * Transforms between CBOR hex string and Native using CBOR encoding.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromCBORHex = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.compose(Schema.Uint8ArrayFromHex,
// string → Uint8Array
FromCBORBytes(options) // Uint8Array → Native
).annotations({
  identifier: "Native.FromCBORHex",
  title: "Native from CBOR Hex",
  description: "Transforms CBOR hex string to Native"
});
/**
 * Root Functions
 * ============================================================================
 */
/**
 * Parse Native from CBOR bytes.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromCBORBytes = /*#__PURE__*/Function.makeCBORDecodeSync(FromCDDL, NativeError, "NativeScripts.fromCBORBytes");
/**
 * Parse Native from CBOR hex string.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromCBORHex = (hex, options) => E.getOrThrow(Either.fromCBORHex(hex, options));
/**
 * Encode Native to CBOR bytes.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toCBORBytes = /*#__PURE__*/Function.makeCBOREncodeSync(FromCDDL, NativeError, "Native.toCBORBytes");
/**
 * Encode Native to CBOR hex string.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toCBORHex = /*#__PURE__*/Function.makeCBOREncodeHexSync(FromCDDL, NativeError, "Native.toCBORHex");
// ============================================================================
// Effect Namespace
// ============================================================================
/**
 * Effect-based error handling variants for functions that can fail.
 *
 * @since 2.0.0
 * @category effect
 */
export var Either;
(function (Either) {
  /**
   * Parse Native from CBOR bytes with Effect error handling.
   *
   * @since 2.0.0
   * @category parsing
   */
  Either.fromCBORBytes = /*#__PURE__*/Function.makeCBORDecodeEither(FromCDDL, NativeError);
  /**
   * Parse Native from CBOR hex string with Effect error handling.
   *
   * @since 2.0.0
   * @category parsing
   */
  Either.fromCBORHex = /*#__PURE__*/Function.makeCBORDecodeHexEither(FromCDDL, NativeError);
  /**
   * Encode Native to CBOR bytes with Effect error handling.
   *
   * @since 2.0.0
   * @category encoding
   */
  Either.toCBORBytes = /*#__PURE__*/Function.makeCBOREncodeEither(FromCDDL, NativeError);
  /**
   * Encode Native to CBOR hex string with Effect error handling.
   *
   * @since 2.0.0
   * @category encoding
   */
  Either.toCBORHex = /*#__PURE__*/Function.makeCBOREncodeHexEither(FromCDDL, NativeError);
})(Either || (Either = {}));
