// @evolution-sdk/evolution/dist/NativeScripts.js
import { Data, Effect as Eff, Equal, FastCheck, Hash, ParseResult, Schema } from "effect";
import * as Bytes from "./Bytes.js";
import * as CBOR from "./CBOR.js";
import * as Hash28 from "./Hash28.js";
/**
 * Error class for Native script related operations.
 *
 * @since 2.0.0
 * @category errors
 */
export class NativeScriptError extends /*#__PURE__*/Data.TaggedError("NativeScriptError") {}
// ============================================================================
// Schema Definition
// ============================================================================
/**
 * Helper function to recursively compare NativeScriptVariants
 */
const equalsVariant = (a, b) => {
  if (a._tag !== b._tag) return false;
  switch (a._tag) {
    case "ScriptPubKey":
      {
        const bScript = b;
        return Bytes.equals(a.keyHash, bScript.keyHash);
      }
    case "InvalidBefore":
      {
        const bScript = b;
        return a.slot === bScript.slot;
      }
    case "InvalidHereafter":
      {
        const bScript = b;
        return a.slot === bScript.slot;
      }
    case "ScriptAll":
      {
        const bScript = b;
        if (a.scripts.length !== bScript.scripts.length) return false;
        for (let i = 0; i < a.scripts.length; i++) {
          if (!equalsVariant(a.scripts[i], bScript.scripts[i])) return false;
        }
        return true;
      }
    case "ScriptAny":
      {
        const bScript = b;
        if (a.scripts.length !== bScript.scripts.length) return false;
        for (let i = 0; i < a.scripts.length; i++) {
          if (!equalsVariant(a.scripts[i], bScript.scripts[i])) return false;
        }
        return true;
      }
    case "ScriptNOfK":
      {
        const bScript = b;
        if (a.required !== bScript.required) return false;
        if (a.scripts.length !== bScript.scripts.length) return false;
        for (let i = 0; i < a.scripts.length; i++) {
          if (!equalsVariant(a.scripts[i], bScript.scripts[i])) return false;
        }
        return true;
      }
  }
};
/**
 * Internal Union schema for the actual native script variants
 *
 * @since 2.0.0
 * @category schemas
 */
export const NativeScriptVariants = /*#__PURE__*/Schema.Union(/*#__PURE__*/Schema.Struct({
  _tag: /*#__PURE__*/Schema.Literal("ScriptPubKey"),
  keyHash: Hash28.BytesFromHex
}), /*#__PURE__*/Schema.Struct({
  _tag: /*#__PURE__*/Schema.Literal("InvalidBefore"),
  slot: Schema.BigInt
}), /*#__PURE__*/Schema.Struct({
  _tag: /*#__PURE__*/Schema.Literal("InvalidHereafter"),
  slot: Schema.BigInt
}), /*#__PURE__*/Schema.Struct({
  _tag: /*#__PURE__*/Schema.Literal("ScriptAll"),
  scripts: /*#__PURE__*/Schema.Array(/*#__PURE__*/Schema.suspend(() => NativeScriptVariants))
}), /*#__PURE__*/Schema.Struct({
  _tag: /*#__PURE__*/Schema.Literal("ScriptAny"),
  scripts: /*#__PURE__*/Schema.Array(/*#__PURE__*/Schema.suspend(() => NativeScriptVariants))
}), /*#__PURE__*/Schema.Struct({
  _tag: /*#__PURE__*/Schema.Literal("ScriptNOfK"),
  required: Schema.BigInt,
  scripts: /*#__PURE__*/Schema.Array(/*#__PURE__*/Schema.suspend(() => NativeScriptVariants))
}));
/**
 * TaggedClass schema for native scripts containing the Union
 *
 * @since 2.0.0
 * @category schemas
 */
export class NativeScript extends /*#__PURE__*/Schema.TaggedClass("NativeScript")("NativeScript", {
  script: NativeScriptVariants
}, {
  identifier: "NativeScript",
  title: "Native Script",
  description: "A native script following Cardano specifications"
}) {
  [Equal.symbol](that) {
    return that instanceof NativeScript && equalsVariant(this.script, that.script);
  }
  [Hash.symbol]() {
    return Hash.hash(JSON.stringify(toJSON(this.script)));
  }
}
// ============================================================================
// Smart Constructors
// ============================================================================
/**
 * Create a signature script for a specific key hash
 *
 * @since 2.0.0
 * @category constructors
 */
export const makeScriptPubKey = keyHash => new NativeScript({
  script: {
    _tag: "ScriptPubKey",
    keyHash
  }
});
/**
 * Create a time-based script that is invalid before a slot
 *
 * @since 2.0.0
 * @category constructors
 */
export const makeInvalidBefore = slot => new NativeScript({
  script: {
    _tag: "InvalidBefore",
    slot
  }
});
/**
 * Create a time-based script that is invalid after a slot
 *
 * @since 2.0.0
 * @category constructors
 */
export const makeInvalidHereafter = slot => new NativeScript({
  script: {
    _tag: "InvalidHereafter",
    slot
  }
});
/**
 * Create a script that requires all nested scripts
 *
 * @since 2.0.0
 * @category constructors
 */
export const makeScriptAll = scripts => new NativeScript({
  script: {
    _tag: "ScriptAll",
    scripts
  }
});
/**
 * Create a script that requires any one nested script
 *
 * @since 2.0.0
 * @category constructors
 */
export const makeScriptAny = scripts => new NativeScript({
  script: {
    _tag: "ScriptAny",
    scripts
  }
});
/**
 * Create a script that requires at least N nested scripts
 *
 * @since 2.0.0
 * @category constructors
 */
export const makeScriptNOfK = (required, scripts) => new NativeScript({
  script: {
    _tag: "ScriptNOfK",
    required,
    scripts
  }
});
// ============================================================================
// JSON Conversion Utilities
// ============================================================================
/**
 * Convert a NativeScript to JSON representation matching cardano-cli format
 *
 * @since 2.0.0
 * @category conversion
 */
export const toJSON = script => {
  switch (script._tag) {
    case "ScriptPubKey":
      return {
        type: "sig",
        keyHash: Bytes.toHex(script.keyHash)
      };
    case "InvalidBefore":
      return {
        type: "after",
        slot: Number(script.slot)
      };
    case "InvalidHereafter":
      return {
        type: "before",
        slot: Number(script.slot)
      };
    case "ScriptAll":
      return {
        type: "all",
        scripts: script.scripts.map(toJSON)
      };
    case "ScriptAny":
      return {
        type: "any",
        scripts: script.scripts.map(toJSON)
      };
    case "ScriptNOfK":
      return {
        type: "atLeast",
        required: Number(script.required),
        scripts: script.scripts.map(toJSON)
      };
  }
};
// Individual CDDL schemas
const ScriptPubKeyCDDL = /*#__PURE__*/Schema.Tuple(/*#__PURE__*/Schema.Literal(0n), Schema.Uint8ArrayFromSelf);
const ScriptAllCDDL = /*#__PURE__*/Schema.Tuple(/*#__PURE__*/Schema.Literal(1n), /*#__PURE__*/Schema.Array(/*#__PURE__*/Schema.suspend(() => CDDLSchema)));
const ScriptAnyCDDL = /*#__PURE__*/Schema.Tuple(/*#__PURE__*/Schema.Literal(2n), /*#__PURE__*/Schema.Array(/*#__PURE__*/Schema.suspend(() => CDDLSchema)));
const ScriptNOfKCDDL = /*#__PURE__*/Schema.Tuple(/*#__PURE__*/Schema.Literal(3n), Schema.BigIntFromSelf, /*#__PURE__*/Schema.Array(/*#__PURE__*/Schema.suspend(() => CDDLSchema)));
const InvalidBeforeCDDL = /*#__PURE__*/Schema.Tuple(/*#__PURE__*/Schema.Literal(4n), Schema.BigIntFromSelf);
const InvalidHereafterCDDL = /*#__PURE__*/Schema.Tuple(/*#__PURE__*/Schema.Literal(5n), Schema.BigIntFromSelf);
export const CDDLSchema = /*#__PURE__*/Schema.Union(ScriptPubKeyCDDL, ScriptAllCDDL, ScriptAnyCDDL, ScriptNOfKCDDL, InvalidBeforeCDDL, InvalidHereafterCDDL).annotations({
  identifier: "NativeScriptCDDL"
});
/**
 * Transform between NativeScript and CDDL representation
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromCDDL = /*#__PURE__*/Schema.transformOrFail(CDDLSchema, Schema.typeSchema(NativeScript), {
  strict: true,
  encode: nativeScript => Eff.gen(function* () {
    const script = nativeScript.script;
    switch (script._tag) {
      case "ScriptPubKey":
        return [0n, script.keyHash];
      case "ScriptAll":
        {
          const encodedScripts = [];
          for (const nestedScript of script.scripts) {
            const nestedNativeScript = new NativeScript({
              script: nestedScript
            });
            const encoded = yield* ParseResult.encode(FromCDDL)(nestedNativeScript);
            encodedScripts.push(encoded);
          }
          return [1n, encodedScripts];
        }
      case "ScriptAny":
        {
          const encodedScripts = [];
          for (const nestedScript of script.scripts) {
            const nestedNativeScript = new NativeScript({
              script: nestedScript
            });
            const encoded = yield* ParseResult.encode(FromCDDL)(nestedNativeScript);
            encodedScripts.push(encoded);
          }
          return [2n, encodedScripts];
        }
      case "ScriptNOfK":
        {
          const encodedScripts = [];
          for (const nestedScript of script.scripts) {
            const nestedNativeScript = new NativeScript({
              script: nestedScript
            });
            const encoded = yield* ParseResult.encode(FromCDDL)(nestedNativeScript);
            encodedScripts.push(encoded);
          }
          return [3n, script.required, encodedScripts];
        }
      case "InvalidBefore":
        return [4n, script.slot];
      case "InvalidHereafter":
        return [5n, script.slot];
    }
  }),
  decode: cddl => Eff.gen(function* () {
    switch (cddl[0]) {
      case 0n:
        {
          const [, keyHash] = cddl;
          return makeScriptPubKey(keyHash);
        }
      case 1n:
        {
          const [, scripts] = cddl;
          const decodedScripts = [];
          for (const scriptCddl of scripts) {
            const decoded = yield* ParseResult.decode(FromCDDL)(scriptCddl);
            decodedScripts.push(decoded.script);
          }
          return makeScriptAll(decodedScripts);
        }
      case 2n:
        {
          const [, scripts] = cddl;
          const decodedScripts = [];
          for (const scriptCddl of scripts) {
            const decoded = yield* ParseResult.decode(FromCDDL)(scriptCddl);
            decodedScripts.push(decoded.script);
          }
          return makeScriptAny(decodedScripts);
        }
      case 3n:
        {
          const [, required, scripts] = cddl;
          const decodedScripts = [];
          for (const scriptCddl of scripts) {
            const decoded = yield* ParseResult.decode(FromCDDL)(scriptCddl);
            decodedScripts.push(decoded.script);
          }
          return makeScriptNOfK(required, decodedScripts);
        }
      case 4n:
        {
          const [, slot] = cddl;
          return makeInvalidBefore(slot);
        }
      case 5n:
        {
          const [, slot] = cddl;
          return makeInvalidHereafter(slot);
        }
      default:
        return yield* ParseResult.fail(new ParseResult.Type(Schema.typeSchema(NativeScript).ast, cddl, `Unknown native script tag: ${cddl[0]}`));
    }
  })
}).annotations({
  identifier: "NativeScript.FromCDDL"
});
export const FromCBORBytes = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.compose(CBOR.FromBytes(options),
// Uint8Array → CBOR
FromCDDL // CBOR → NativeScript
);
export const FromCBORHex = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.compose(Schema.Uint8ArrayFromHex,
// string → Uint8Array
FromCBORBytes(options) // Uint8Array → NativeScript
);
/**
 * Check if the given value is a valid NativeScript
 *
 * @since 2.0.0
 * @category predicates
 */
export const is = /*#__PURE__*/Schema.is(NativeScriptVariants);
/**
 * FastCheck arbitrary for generating random NativeScript instances
 *
 * @since 2.0.0
 * @category testing
 */
export const arbitrary = /*#__PURE__*/FastCheck.letrec(tie => ({
  nativeScript: FastCheck.oneof(
  // ScriptPubKey
  FastCheck.uint8Array({
    minLength: 28,
    maxLength: 28
  }).map(keyHash => makeScriptPubKey(keyHash)),
  // InvalidBefore
  FastCheck.bigInt({
    min: 0n,
    max: 2n ** 64n - 1n
  }).map(slot => makeInvalidBefore(slot)),
  // InvalidHereafter
  FastCheck.bigInt({
    min: 0n,
    max: 2n ** 64n - 1n
  }).map(slot => makeInvalidHereafter(slot)),
  // ScriptAll (limit depth to prevent infinite recursion)
  FastCheck.array(tie("nativeScriptVariant"), {
    maxLength: 3
  }).map(scripts => makeScriptAll(scripts)),
  // ScriptAny (limit depth to prevent infinite recursion)
  FastCheck.array(tie("nativeScriptVariant"), {
    maxLength: 3
  }).map(scripts => makeScriptAny(scripts)),
  // ScriptNOfK (limit depth to prevent infinite recursion)
  FastCheck.tuple(FastCheck.bigInt({
    min: 0n,
    max: 10n
  }), FastCheck.array(tie("nativeScriptVariant"), {
    maxLength: 3
  })).map(([required, scripts]) => makeScriptNOfK(required, scripts))),
  // IMPORTANT: this generates NativeScriptVariants (plain variant objects), not NativeScript wrappers
  nativeScriptVariant: FastCheck.oneof(
  // ScriptPubKey
  FastCheck.uint8Array({
    minLength: 28,
    maxLength: 28
  }).map(keyHash => ({
    _tag: "ScriptPubKey",
    keyHash
  })),
  // InvalidBefore
  FastCheck.bigInt({
    min: 0n,
    max: 2n ** 64n - 1n
  }).map(slot => ({
    _tag: "InvalidBefore",
    slot
  })),
  // InvalidHereafter
  FastCheck.bigInt({
    min: 0n,
    max: 2n ** 64n - 1n
  }).map(slot => ({
    _tag: "InvalidHereafter",
    slot
  })),
  // ScriptAll (limit depth to prevent infinite recursion)
  FastCheck.array(tie("nativeScriptVariant"), {
    maxLength: 2
  }).map(scripts => ({
    _tag: "ScriptAll",
    scripts: scripts
  })),
  // ScriptAny (limit depth to prevent infinite recursion)
  FastCheck.array(tie("nativeScriptVariant"), {
    maxLength: 2
  }).map(scripts => ({
    _tag: "ScriptAny",
    scripts: scripts
  })),
  // ScriptNOfK (limit depth to prevent infinite recursion)
  FastCheck.tuple(FastCheck.bigInt({
    min: 0n,
    max: 10n
  }), FastCheck.array(tie("nativeScriptVariant"), {
    maxLength: 2
  })).map(([required, scripts]) => ({
    _tag: "ScriptNOfK",
    required,
    scripts: scripts
  })))
})).nativeScript;
/**
 * Parse a NativeScript from CBOR bytes.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromCBORBytes = (bytes, options) => Schema.decodeSync(FromCBORBytes(options))(bytes);
/**
 * Parse a NativeScript from CBOR hex string.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromCBORHex = (hex, options) => Schema.decodeSync(FromCBORHex(options))(hex);
/**
 * Convert a NativeScript to CBOR bytes.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toCBORBytes = (nativeScript, options) => Schema.encodeSync(FromCBORBytes(options))(nativeScript);
/**
 * Convert a NativeScript to CBOR hex string.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toCBORHex = (nativeScript, options) => Schema.encodeSync(FromCBORHex(options))(nativeScript);
// ============================================================================
// Required Signers Calculation
// ============================================================================
/**
 * Count the maximum number of key hashes (signers) required to satisfy a native script.
 * This is used for fee calculation to ensure the fake witness set has the correct size.
 *
 * Algorithm:
 * - ScriptPubKey: 1 signer required
 * - ScriptAll: sum of all nested scripts (all must be satisfied)
 * - ScriptAny: maximum of nested scripts (pessimistic - assume most expensive path)
 * - ScriptNOfK: sum of top N most expensive nested scripts
 * - InvalidBefore/InvalidHereafter: 0 signers (time-based only)
 *
 * @since 2.0.0
 * @category utilities
 */
export const countRequiredSigners = script => {
  switch (script._tag) {
    case "ScriptPubKey":
      return 1;
    case "InvalidBefore":
    case "InvalidHereafter":
      return 0;
    case "ScriptAll":
      {
        // All nested scripts must be satisfied, sum them all
        return script.scripts.reduce((sum, nested) => sum + countRequiredSigners(nested), 0);
      }
    case "ScriptAny":
      {
        // Any one nested script can satisfy, use maximum (pessimistic for fee calculation)
        if (script.scripts.length === 0) return 0;
        return Math.max(...script.scripts.map(countRequiredSigners));
      }
    case "ScriptNOfK":
      {
        // At least N scripts must be satisfied
        // Pessimistic approach: assume we need the N most expensive paths
        if (script.scripts.length === 0) return 0;
        const counts = script.scripts.map(countRequiredSigners).sort((a, b) => b - a);
        const required = Number(script.required);
        return counts.slice(0, required).reduce((sum, count) => sum + count, 0);
      }
    default:
      return 0;
  }
};
/**
 * Extract all key hashes from a native script.
 * Recursively traverses nested scripts to find all ScriptPubKey key hashes.
 *
 * @since 2.0.0
 * @category utilities
 */
export const extractKeyHashes = script => {
  const keyHashes = [];
  const traverse = s => {
    switch (s._tag) {
      case "ScriptPubKey":
        keyHashes.push(s.keyHash);
        break;
      case "ScriptAll":
      case "ScriptAny":
        for (const nested of s.scripts) traverse(nested);
        break;
      case "ScriptNOfK":
        for (const nested of s.scripts) traverse(nested);
        break;
      case "InvalidBefore":
      case "InvalidHereafter":
        // Time-based scripts don't contain key hashes
        break;
    }
  };
  traverse(script);
  return keyHashes;
};
