// @evolution-sdk/evolution/dist/NonnegativeInterval.js
import { Effect, Equal, FastCheck, Hash, Inspectable, ParseResult, Schema } from "effect";
import * as CBOR from "./CBOR.js";
import * as Numeric from "./Numeric.js";
/**
 * Schema for NonnegativeInterval representing a fractional value >= 0.
 *
 * CDDL: nonnegative_interval = #6.30([uint, positive_int])
 *
 * @since 2.0.0
 * @category model
 */
export class NonnegativeInterval extends /*#__PURE__*/Schema.Class("NonnegativeInterval")({
  numerator: Numeric.Uint64Schema,
  denominator: Numeric.Uint64Schema // positive_int (we enforce > 0 below)
}) {
  /**
   * Convert to JSON representation.
   *
   * @since 2.0.0
   * @category conversions
   */
  toJSON() {
    return {
      _tag: "NonnegativeInterval",
      numerator: this.numerator,
      denominator: this.denominator
    };
  }
  /**
   * Convert to string representation.
   *
   * @since 2.0.0
   * @category conversions
   */
  toString() {
    return Inspectable.format(this.toJSON());
  }
  /**
   * Custom inspect for Node.js REPL.
   *
   * @since 2.0.0
   * @category conversions
   */
  [Inspectable.NodeInspectSymbol]() {
    return this.toJSON();
  }
  /**
   * Structural equality check.
   *
   * @since 2.0.0
   * @category equality
   */
  [Equal.symbol](that) {
    return that instanceof NonnegativeInterval && this.numerator === that.numerator && this.denominator === that.denominator;
  }
  /**
   * Hash code generation.
   *
   * @since 2.0.0
   * @category hashing
   */
  [Hash.symbol]() {
    return Hash.cached(this, Hash.combine(Hash.hash(this.numerator))(Hash.hash(this.denominator)));
  }
  /**
   * Static filter for validation.
   *
   * @since 2.0.0
   * @category validation
   */
  static get schema() {
    return Schema.filter(interval => {
      if (interval.denominator <= 0n) {
        return {
          path: ["denominator"],
          message: `denominator (${interval.denominator}) must be > 0`
        };
      }
      return true;
    })(NonnegativeInterval);
  }
}
export const CDDLSchema = /*#__PURE__*/CBOR.tag(30, /*#__PURE__*/Schema.Tuple(CBOR.Integer, CBOR.Integer));
/**
 * Transform between tag(30) tuple and NonnegativeInterval model.
 */
export const FromCDDL = /*#__PURE__*/Schema.transformOrFail(CDDLSchema, Schema.typeSchema(NonnegativeInterval), {
  strict: true,
  encode: (_, __, ___, interval) => Effect.succeed({
    _tag: "Tag",
    tag: 30,
    value: [interval.numerator, interval.denominator]
  }),
  decode: (_, __, ___, taggedValue) => Effect.gen(function* () {
    if (taggedValue.tag !== 30) {
      return yield* Effect.fail(new ParseResult.Type(NonnegativeInterval.ast, taggedValue, `Expected tag 30, got ${taggedValue.tag}`));
    }
    const [numerator, denominator] = yield* ParseResult.decodeUnknown(Schema.Tuple(CBOR.Integer, CBOR.Integer))(taggedValue.value);
    return new NonnegativeInterval({
      numerator,
      denominator
    });
  })
}).annotations({
  identifier: "NonnegativeInterval.CDDL"
});
export const FromCBORBytes = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.compose(CBOR.FromBytes(options), FromCDDL);
export const FromCBORHex = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.compose(Schema.Uint8ArrayFromHex, FromCBORBytes(options));
export const arbitrary = /*#__PURE__*/FastCheck.bigInt({
  min: 1n,
  max: 1000000n
}).chain(denominator => FastCheck.bigInt({
  min: 0n,
  max: denominator
}).map(numerator => ({
  numerator,
  denominator
}))).map(v => new NonnegativeInterval(v));
