// @evolution-sdk/evolution/dist/Pointer.js
import { Equal, FastCheck, Hash, Inspectable, Schema } from "effect";
import * as Natural from "./Natural.js";
/**
 * Schema for pointer to a stake registration certificate
 * Contains slot, transaction index, and certificate index information
 *
 * @since 2.0.0
 * @category schemas
 */
export class Pointer extends /*#__PURE__*/Schema.TaggedClass("Pointer")("Pointer", {
  slot: Natural.Natural,
  txIndex: Natural.Natural,
  certIndex: Natural.Natural
}) {
  toJSON() {
    return {
      _tag: "Pointer",
      slot: this.slot,
      txIndex: this.txIndex,
      certIndex: this.certIndex
    };
  }
  toString() {
    return Inspectable.format(this.toJSON());
  }
  [Inspectable.NodeInspectSymbol]() {
    return this.toJSON();
  }
  [Equal.symbol](that) {
    return that instanceof Pointer && Equal.equals(this.slot, that.slot) && Equal.equals(this.txIndex, that.txIndex) && Equal.equals(this.certIndex, that.certIndex);
  }
  [Hash.symbol]() {
    return Hash.combine(Hash.hash(this.slot))(Hash.combine(Hash.hash(this.txIndex))(Hash.hash(this.certIndex)));
  }
}
/**
 * Check if the given value is a valid Pointer
 *
 *
 * @since 2.0.0
 * @category predicates
 */
export const isPointer = /*#__PURE__*/Schema.is(Pointer);
/**
 * FastCheck arbitrary for generating random Pointer instances
 *
 * @since 2.0.0
 * @category generators
 */
export const arbitrary = /*#__PURE__*/FastCheck.tuple(Natural.arbitrary, Natural.arbitrary, Natural.arbitrary).map(([slot, txIndex, certIndex]) => new Pointer({
  slot,
  txIndex,
  certIndex
}, {
  disableValidation: true
}));
