// @evolution-sdk/evolution/dist/PoolKeyHash.js
import { bech32 } from "@scure/base";
import { Effect as Eff, Equal, FastCheck, Hash, Inspectable, ParseResult, Schema } from "effect";
import * as Bytes from "./Bytes.js";
import * as Hash28 from "./Hash28.js";
/**
 * PoolKeyHash as a TaggedClass representing a stake pool's verification key hash.
 * pool_keyhash = hash28
 *
 * @since 2.0.0
 * @category model
 */
export class PoolKeyHash extends /*#__PURE__*/Schema.TaggedClass()("PoolKeyHash", {
  hash: Hash28.BytesFromHex
}) {
  toJSON() {
    return {
      _tag: "PoolKeyHash",
      hash: Bytes.toHex(this.hash)
    };
  }
  toString() {
    return Inspectable.format(this.toJSON());
  }
  [Inspectable.NodeInspectSymbol]() {
    return this.toJSON();
  }
  [Equal.symbol](that) {
    return that instanceof PoolKeyHash && Bytes.equals(this.hash, that.hash);
  }
  [Hash.symbol]() {
    return Hash.cached(this, Hash.array(Array.from(this.hash)));
  }
}
/**
 * Schema transformer from bytes to PoolKeyHash.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromBytes = /*#__PURE__*/Schema.transform(Schema.typeSchema(Hash28.BytesFromHex), Schema.typeSchema(PoolKeyHash), {
  strict: true,
  decode: hash => new PoolKeyHash({
    hash
  }, {
    disableValidation: true
  }),
  encode: poolKeyHash => poolKeyHash.hash
}).annotations({
  identifier: "PoolKeyHash.FromBytes"
});
/**
 * Schema transformer from hex string to PoolKeyHash.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromHex = /*#__PURE__*/Schema.compose(Hash28.BytesFromHex, FromBytes).annotations({
  identifier: "PoolKeyHash.FromHex"
});
/**
 * Schema transformer from bech32 string (pool1...) to PoolKeyHash.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromBech32 = /*#__PURE__*/Schema.transformOrFail(Schema.String, Schema.typeSchema(PoolKeyHash), {
  strict: true,
  encode: poolKeyHash => Eff.gen(function* () {
    const words = bech32.toWords(poolKeyHash.hash);
    return bech32.encode("pool", words, false);
  }),
  decode: (fromA, _, ast) => Eff.gen(function* () {
    const result = yield* Eff.try({
      try: () => {
        // Note: `as any` needed because bech32.decode expects template literal type `${Prefix}1${string}`
        // but Schema provides plain string. Consider using decodeToBytes which accepts string.
        const decoded = bech32.decode(fromA, false);
        const bytes = bech32.fromWords(decoded.words);
        return new Uint8Array(bytes);
      },
      catch: () => new ParseResult.Type(ast, fromA, `Failed to decode Bech32 pool id: ${fromA}`)
    });
    return yield* ParseResult.decode(FromBytes)(result);
  })
}).annotations({
  identifier: "PoolKeyHash.FromBech32",
  description: "Transforms Bech32 pool id string to PoolKeyHash"
});
/**
 * FastCheck arbitrary for generating random PoolKeyHash instances.
 *
 * @since 2.0.0
 * @category arbitrary
 */
export const arbitrary = /*#__PURE__*/FastCheck.uint8Array({
  minLength: Hash28.BYTES_LENGTH,
  maxLength: Hash28.BYTES_LENGTH
}).map(bytes => new PoolKeyHash({
  hash: bytes
}, {
  disableValidation: true
}));
// ============================================================================
// Root Functions
// ============================================================================
/**
 * Parse PoolKeyHash from bytes.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromBytes = /*#__PURE__*/Schema.decodeSync(FromBytes);
/**
 * Parse PoolKeyHash from hex string.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromHex = /*#__PURE__*/Schema.decodeSync(FromHex);
/**
 * Encode PoolKeyHash to bytes.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toBytes = /*#__PURE__*/Schema.encodeSync(FromBytes);
/**
 * Encode PoolKeyHash to hex string.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toHex = /*#__PURE__*/Schema.encodeSync(FromHex);
/**
 * Parse PoolKeyHash from bech32 string (pool1...).
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromBech32 = /*#__PURE__*/Schema.decodeSync(FromBech32);
/**
 * Encode PoolKeyHash to bech32 string (pool1...).
 *
 * @since 2.0.0
 * @category encoding
 */
export const toBech32 = /*#__PURE__*/Schema.encodeSync(FromBech32);
