// @evolution-sdk/evolution/dist/Port.js
import { Schema } from "effect";
import * as Numeric from "./Numeric.js";
/**
 * Schema for validating port numbers (0-65535).
 *
 * @since 2.0.0
 * @category schemas
 */
export const PortSchema = /*#__PURE__*/Numeric.Uint16Schema.annotations({
  identifier: "Port",
  description: "Network port number (16-bit unsigned integer)"
});
/**
 * Check if a value is a valid Port.
 *
 * @since 2.0.0
 * @category predicates
 */
export const is = value => Schema.is(PortSchema)(value);
/**
 * Check if a port is a well-known port (0-1023).
 *
 * @since 2.0.0
 * @category predicates
 */
export const isWellKnown = port => port >= 0 && port <= 1023;
/**
 * Check if a port is a registered port (1024-49151).
 *
 * @since 2.0.0
 * @category predicates
 */
export const isRegistered = port => port >= 1024 && port <= 49151;
/**
 * Check if a port is a dynamic/private port (49152-65535).
 *
 * @since 2.0.0
 * @category predicates
 */
export const isDynamic = port => port >= 49152 && port <= 65535;
/**
 * Generate a random Port.
 *
 * @since 2.0.0
 * @category arbitrary
 */
export const arbitrary = Numeric.Uint16Arbitrary;
