// @evolution-sdk/evolution/dist/Relay.js
import { FastCheck, Schema } from "effect";
import * as CBOR from "./CBOR.js";
import * as MultiHostName from "./MultiHostName.js";
import * as SingleHostAddr from "./SingleHostAddr.js";
import * as SingleHostName from "./SingleHostName.js";
/**
 * Union schema for Relay representing various relay configurations.
 * relay = [ single_host_addr // single_host_name // multi_host_name ]
 *
 * @since 2.0.0
 * @category schemas
 */
export const Relay = /*#__PURE__*/Schema.Union(SingleHostAddr.SingleHostAddr, SingleHostName.SingleHostName, MultiHostName.MultiHostName);
export const FromCDDL = /*#__PURE__*/Schema.Union(SingleHostAddr.FromCDDL, SingleHostName.FromCDDL, MultiHostName.FromCDDL);
/**
 * CBOR bytes transformation schema for Relay.
 * For union types, we create a union of the child CBOR schemas.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromCBORBytes = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.Union(SingleHostAddr.FromCBORBytes(options), SingleHostName.FromBytes(options),
// Still uses old naming
MultiHostName.FromCBORBytes(options)).annotations({
  identifier: "Relay.FromCBORBytes",
  title: "Relay from CBOR Bytes",
  description: "Transforms CBOR bytes (Uint8Array) to Relay"
});
/**
 * CBOR hex transformation schema for Relay.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromCBORHex = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.compose(Schema.Uint8ArrayFromHex,
// string → Uint8Array
FromCBORBytes(options) // Uint8Array → Relay
).annotations({
  identifier: "Relay.FromCBORHex",
  title: "Relay from CBOR Hex",
  description: "Transforms CBOR hex string to Relay"
});
/**
 * @since 2.0.0
 * @category FastCheck
 */
export const arbitrary = /*#__PURE__*/FastCheck.oneof(SingleHostAddr.arbitrary, SingleHostName.arbitrary, MultiHostName.arbitrary);
/**
 * Create a Relay from a SingleHostAddr.
 *
 * @since 2.0.0
 * @category constructors
 */
export const fromSingleHostAddr = singleHostAddr => singleHostAddr;
/**
 * Create a Relay from a SingleHostName.
 *
 * @since 2.0.0
 * @category constructors
 */
export const fromSingleHostName = singleHostName => singleHostName;
/**
 * Create a Relay from a MultiHostName.
 *
 * @since 2.0.0
 * @category constructors
 */
export const fromMultiHostName = multiHostName => multiHostName;
/**
 * Parse Relay from CBOR bytes.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromCBORBytes = (bytes, options) => Schema.decodeSync(FromCBORBytes(options))(bytes);
/**
 * Parse Relay from CBOR hex.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromCBORHex = (hex, options) => Schema.decodeSync(FromCBORHex(options))(hex);
/**
 * Convert Relay to CBOR bytes.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toCBORBytes = (data, options) => Schema.encodeSync(FromCBORBytes(options))(data);
/**
 * Convert Relay to CBOR hex.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toCBORHex = (data, options) => Schema.encodeSync(FromCBORHex(options))(data);
/**
 * Pattern match on a Relay to handle different relay types.
 *
 * @since 2.0.0
 * @category transformation
 */
export const match = (relay, cases) => {
  switch (relay._tag) {
    case "SingleHostAddr":
      return cases.SingleHostAddr(relay);
    case "SingleHostName":
      return cases.SingleHostName(relay);
    case "MultiHostName":
      return cases.MultiHostName(relay);
    default:
      throw new Error(`Exhaustive check failed: Unhandled case '${relay._tag}' encountered.`);
  }
};
/**
 * Check if a Relay is a SingleHostAddr.
 *
 * @since 2.0.0
 * @category predicates
 */
export const isSingleHostAddr = relay => relay._tag === "SingleHostAddr";
/**
 * Check if a Relay is a SingleHostName.
 *
 * @since 2.0.0
 * @category predicates
 */
export const isSingleHostName = relay => relay._tag === "SingleHostName";
/**
 * Check if a Relay is a MultiHostName.
 *
 * @since 2.0.0
 * @category predicates
 */
export const isMultiHostName = relay => relay._tag === "MultiHostName";
