// @evolution-sdk/evolution/dist/SingleHostName.js
import { Effect, Equal, FastCheck, Hash, Inspectable, ParseResult, Schema } from "effect";
import * as CBOR from "./CBOR.js";
import * as DnsName from "./DnsName.js";
import * as Port from "./Port.js";
/**
 * Schema for SingleHostName representing a network host with DNS name.
 * single_host_name = (1, port/ nil, dns_name)
 *
 * Used for A or AAAA DNS records.
 *
 * @since 2.0.0
 * @category model
 */
export class SingleHostName extends /*#__PURE__*/Schema.TaggedClass()("SingleHostName", {
  port: /*#__PURE__*/Schema.optional(Port.PortSchema),
  dnsName: DnsName.DnsName
}) {
  /**
   * Convert to JSON-serializable format.
   * Relies on Option's built-in toJSON() for port serialization.
   * Converts bigint port values to strings for JSON compatibility.
   *
   * @since 2.0.0
   * @category serialization
   */
  toJSON() {
    return {
      _tag: "SingleHostName",
      port: this.port !== undefined ? String(this.port) : null,
      dnsName: this.dnsName
    };
  }
  toString() {
    return Inspectable.format(this.toJSON());
  }
  [Inspectable.NodeInspectSymbol]() {
    return this.toJSON();
  }
  [Equal.symbol](that) {
    return that instanceof SingleHostName && this.port === that.port && Equal.equals(this.dnsName, that.dnsName);
  }
  [Hash.symbol]() {
    return Hash.cached(this, Hash.combine(Hash.hash(this.port))(Hash.hash(this.dnsName)));
  }
  /**
   * Convert to CBOR bytes.
   *
   * @since 2.0.0
   * @category serialization
   */
  toCBORBytes() {
    return toCBORBytes(this);
  }
  /**
   * Convert to CBOR hex string.
   *
   * @since 2.0.0
   * @category serialization
   */
  toCBORHex() {
    return toCBORHex(this);
  }
  [Symbol.for("nodejs.util.inspect.custom")]() {
    return {
      _tag: "SingleHostName",
      port: this.port,
      dnsName: this.dnsName
    };
  }
}
/**
 * Create a SingleHostName with a port.
 *
 * @since 2.0.0
 * @category constructors
 */
export const withPort = (port, dnsName) => new SingleHostName({
  port,
  dnsName
});
/**
 * Create a SingleHostName without a port.
 *
 * @since 2.0.0
 * @category constructors
 */
export const withoutPort = dnsName => new SingleHostName({
  dnsName
});
/**
 * Check if the host name has a port.
 *
 * @since 2.0.0
 * @category predicates
 */
export const hasPort = hostName => hostName.port !== undefined;
/**
 * Get the DNS name from a SingleHostName.
 *
 * @since 2.0.0
 * @category transformation
 */
export const getDnsName = hostName => hostName.dnsName;
/**
 * Get the port from a SingleHostName, if it exists.
 *
 * @since 2.0.0
 * @category transformation
 */
export const getPort = hostName => hostName.port;
/**
 * Generate a random SingleHostName.
 *
 * @since 2.0.0
 * @category generators
 */
export const generator = /*#__PURE__*/FastCheck.record({
  port: FastCheck.option(Port.arbitrary),
  dnsName: DnsName.arbitrary
}).map(({
  dnsName,
  port
}) => new SingleHostName({
  port: port === null ? undefined : port,
  dnsName
}));
/**
 * FastCheck arbitrary for SingleHostName instances.
 * Alias to `generator` for consistency with other modules.
 *
 * @since 2.0.0
 * @category testing
 */
export const arbitrary = generator;
/**
 * CDDL schema for SingleHostName.
 * single_host_name = (1, port / nil, dns_name)
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromCDDL = /*#__PURE__*/Schema.transformOrFail(/*#__PURE__*/Schema.Tuple(/*#__PURE__*/Schema.Literal(1n),
/*#__PURE__*/
// tag (literal 1)
Schema.NullOr(CBOR.Integer),
// port (number or null)
Schema.String // dns_name (string)
), /*#__PURE__*/Schema.typeSchema(SingleHostName), {
  strict: true,
  encode: toA => Effect.gen(function* () {
    const port = toA.port !== undefined ? toA.port : null;
    const dnsName = yield* ParseResult.encode(DnsName.DnsName)(toA.dnsName);
    return yield* Effect.succeed([1n, port, dnsName]);
  }),
  decode: ([, portValue, dnsNameValue]) => Effect.gen(function* () {
    const port = portValue === null || portValue === undefined ? undefined : portValue;
    const dnsName = yield* ParseResult.decode(DnsName.DnsName)(dnsNameValue);
    return new SingleHostName({
      port,
      dnsName
    });
  })
});
/**
 * CBOR bytes transformation schema for SingleHostName.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromBytes = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.compose(CBOR.FromBytes(options),
// Uint8Array → CBOR
FromCDDL // CBOR → SingleHostName
);
/**
 * CBOR hex transformation schema for SingleHostName.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromHex = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.compose(Schema.Uint8ArrayFromHex,
// string → Uint8Array
FromBytes(options) // Uint8Array → SingleHostName
);
// ============================================================================
// Root Functions
// ============================================================================
/**
 * Parse a SingleHostName from CBOR bytes.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromCBORBytes = (bytes, options) => Schema.decodeSync(FromBytes(options))(bytes);
/**
 * Parse a SingleHostName from CBOR hex string.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromCBORHex = (hex, options) => Schema.decodeSync(FromHex(options))(hex);
// ============================================================================
// Encoding Functions
// ============================================================================
/**
 * Convert a SingleHostName to CBOR bytes.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toCBORBytes = (data, options) => Schema.encodeSync(FromBytes(options))(data);
/**
 * Convert a SingleHostName to CBOR hex string.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toCBORHex = (data, options) => Schema.encodeSync(FromHex(options))(data);
