// @evolution-sdk/evolution/dist/Text.js
import { FastCheck, Schema } from "effect";
export const Text = Schema.String;
// Shared transform options for text conversions
const textTransform = {
  strict: true,
  decode: input => new TextDecoder().decode(input),
  encode: text => new TextEncoder().encode(text)
};
/**
 * Schema for converting between strings and bytes (UTF-8).
 *
 * ```
 * text <-> bytes
 * ```
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromBytes = /*#__PURE__*/Schema.transform(Schema.Uint8ArrayFromSelf, Text, textTransform).annotations({
  identifier: "Text.FromBytes"
});
/**
 * Schema for converting between strings and hex representation of UTF-8 bytes.
 *
 * ```
 * text <-> hex
 * ```
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromHex = /*#__PURE__*/Schema.transform(Schema.Uint8ArrayFromHex, Text, textTransform).annotations({
  identifier: "Text.FromHex"
});
// =============================================================================
// Text Length Validation Utilities
// =============================================================================
/**
 * Creates a schema that validates text length equals a specific value.
 *
 * @since 2.0.0
 * @category validation
 */
export const length = (length, identifier) => Schema.filter(text => text.length === length, {
  message: () => `Expected text length ${length}`,
  identifier
});
/**
 * Creates a curried filter that validates text length is within a range.
 * Preserves Context inference from the base schema.
 *
 * @since 2.0.0
 * @category composition
 */
export const between = (min, max, moduleName) => baseSchema => baseSchema.pipe(Schema.filter(text => {
  const textLength = text.length;
  return textLength >= min && textLength <= max;
}, {
  message: () => `Must be between ${min} and ${max} characters`,
  identifier: `${moduleName}.LengthBetween${min}And${max}`
}));
/**
 * Creates a schema that validates text length is at least min.
 *
 * @since 2.0.0
 * @category validation
 */
export const min = (min, identifier) => Schema.filter(text => text.length >= min, {
  message: () => `Expected text length at least ${min}`,
  identifier
});
/**
 * Creates a schema that validates text length is at most max.
 *
 * @since 2.0.0
 * @category validation
 */
export const max = (max, identifier) => Schema.filter(text => text.length <= max, {
  message: () => `Expected text length at most ${max}`,
  identifier
});
/**
 * FastCheck arbitrary for generating random text strings
 *
 * @since 2.0.0
 * @category testing
 */
export const arbitrary = /*#__PURE__*/FastCheck.string();
// =============================================================================
// Public (throwing) API
// =============================================================================
/**
 * Convert bytes to text
 *
 * @since 2.0.0
 * @category conversion
 */
export const fromBytes = /*#__PURE__*/Schema.decodeSync(FromBytes);
/**
 * Convert hex string to text
 *
 * @since 2.0.0
 * @category conversion
 */
export const fromHex = /*#__PURE__*/Schema.decodeSync(FromHex);
/**
 * Convert text to bytes
 *
 * @since 2.0.0
 * @category conversion
 */
export const toBytes = /*#__PURE__*/Schema.encodeSync(FromBytes);
/**
 * Convert text to hex string
 *
 * @since 2.0.0
 * @category conversion
 */
export const toHex = /*#__PURE__*/Schema.encodeSync(FromHex);
