// @evolution-sdk/evolution/dist/TransactionOutput.js
import { Either as E, Equal, FastCheck, Hash, Inspectable, ParseResult, Schema } from "effect";
import * as AddressEras from "./AddressEras.js";
import * as BaseAddress from "./BaseAddress.js";
import * as CBOR from "./CBOR.js";
import * as DatumHash from "./DatumHash.js";
import * as DatumOption from "./DatumOption.js";
import * as EnterpriseAddress from "./EnterpriseAddress.js";
import * as ScriptRef from "./ScriptRef.js";
import * as Value from "./Value.js";
// Pre-bind frequently used ParseResult helpers for hot paths
const encAddress = /*#__PURE__*/ParseResult.encodeEither(AddressEras.FromBytes);
const decAddress = /*#__PURE__*/ParseResult.decodeUnknownEither(/*#__PURE__*/Schema.Union(BaseAddress.FromBytes, EnterpriseAddress.FromBytes));
const encValue = /*#__PURE__*/ParseResult.encodeEither(Value.FromCDDL);
const decValue = /*#__PURE__*/ParseResult.decodeUnknownEither(Value.FromCDDL);
const encDatumOption = /*#__PURE__*/ParseResult.encodeEither(DatumOption.FromCDDL);
const decDatumOption = /*#__PURE__*/ParseResult.decodeUnknownEither(DatumOption.FromCDDL);
const decDatumHash = /*#__PURE__*/ParseResult.decodeEither(DatumHash.FromBytes);
const encScriptRef = /*#__PURE__*/ParseResult.encodeEither(ScriptRef.FromCDDL);
const decScriptRef = /*#__PURE__*/ParseResult.decodeUnknownEither(ScriptRef.FromCDDL);
/**
 * Shelley-era transaction output format
 *
 * CDDL:
 * ```
 * shelley_transaction_output = [address, amount : value, ? Bytes32]
 * ```
 *
 * @since 2.0.0
 * @category model
 */
export class ShelleyTransactionOutput extends /*#__PURE__*/Schema.TaggedClass()("ShelleyTransactionOutput", {
  address: AddressEras.FromBech32,
  // Schema.Union(BaseAddress.BaseAddress, EnterpriseAddress.EnterpriseAddress),
  amount: Value.Value,
  datumHash: /*#__PURE__*/Schema.optional(DatumHash.DatumHash)
}) {
  toJSON() {
    return {
      _tag: this._tag,
      address: this.address,
      amount: this.amount,
      datumHash: this.datumHash
    };
  }
  toString() {
    return Inspectable.format(this.toJSON());
  }
  [Inspectable.NodeInspectSymbol]() {
    return this.toJSON();
  }
  [Equal.symbol](that) {
    return that instanceof ShelleyTransactionOutput && Equal.equals(this.address, that.address) && Equal.equals(this.amount, that.amount) && Equal.equals(this.datumHash, that.datumHash);
  }
  [Hash.symbol]() {
    return Hash.cached(this, Hash.hash(this.address) ^ Hash.hash(this.amount) ^ Hash.hash(this.datumHash));
  }
}
/**
 * Babbage-era transaction output format
 *
 * CDDL:
 * ```
 * babbage_transaction_output =
 *   {0 : address, 1 : value, ? 2 : datum_option, ? 3 : script_ref}
 * ```
 *
 * @since 2.0.0
 * @category model
 */
export class BabbageTransactionOutput extends /*#__PURE__*/Schema.TaggedClass()("BabbageTransactionOutput", {
  address: AddressEras.FromBech32,
  amount: Value.Value,
  // 1
  datumOption: /*#__PURE__*/Schema.optional(DatumOption.DatumOptionSchema),
  // 2
  scriptRef: /*#__PURE__*/Schema.optional(ScriptRef.ScriptRef) // 3
}) {
  toJSON() {
    return {
      _tag: this._tag,
      address: this.address,
      amount: this.amount,
      datumOption: this.datumOption,
      scriptRef: this.scriptRef
    };
  }
  toString() {
    return Inspectable.format(this.toJSON());
  }
  [Inspectable.NodeInspectSymbol]() {
    return this.toJSON();
  }
  [Equal.symbol](that) {
    return that instanceof BabbageTransactionOutput && Equal.equals(this.address, that.address) && Equal.equals(this.amount, that.amount) && Equal.equals(this.datumOption, that.datumOption) && Equal.equals(this.scriptRef, that.scriptRef);
  }
  [Hash.symbol]() {
    return Hash.cached(this, Hash.hash(this.address) ^ Hash.hash(this.amount) ^ Hash.hash(this.datumOption) ^ Hash.hash(this.scriptRef));
  }
}
/**
 * Union type for transaction outputs
 *
 * CDDL:
 * ```
 * transaction_output = shelley_transaction_output / babbage_transaction_output
 * ```
 *
 * @since 2.0.0
 * @category schemas
 */
export const TransactionOutput = /*#__PURE__*/Schema.Union(ShelleyTransactionOutput, BabbageTransactionOutput);
export const ShelleyTransactionOutputCDDL = /*#__PURE__*/Schema.Tuple(Schema.Uint8ArrayFromSelf,
// address as bytes
Value.CDDLSchema,
/*#__PURE__*/
// value
Schema.optionalElement(Schema.Uint8ArrayFromSelf) // optional datum_hash as bytes
);
/**
 * CDDL schema for Shelley transaction outputs
 *
 * @since 2.0.0
 * @category transformation
 */
export const FromShelleyTransactionOutputCDDLSchema = /*#__PURE__*/Schema.transformOrFail(ShelleyTransactionOutputCDDL, /*#__PURE__*/Schema.typeSchema(ShelleyTransactionOutput), {
  strict: true,
  encode: toI => E.gen(function* () {
    const addressBytes = yield* encAddress(toI.address);
    const valueBytes = yield* encValue(toI.amount);
    if (toI.datumHash !== undefined) {
      return [addressBytes, valueBytes, toI.datumHash.hash];
    }
    return [addressBytes, valueBytes];
  }),
  decode: fromI => E.gen(function* () {
    const [addressBytes, valueBytes, datumHashBytes] = fromI;
    const address = yield* decAddress(addressBytes);
    const amount = yield* decValue(valueBytes);
    let datumHash;
    if (datumHashBytes !== undefined) {
      datumHash = yield* decDatumHash(datumHashBytes);
    }
    return new ShelleyTransactionOutput({
      address,
      amount,
      datumHash
    }, {
      disableValidation: true
    });
  })
});
const BabbageTransactionOutputCDDL = /*#__PURE__*/Schema.MapFromSelf({
  key: CBOR.Integer,
  value: CBOR.CBORSchema
});
/**
 * CDDL schema for Babbage transaction outputs
 *
 * @since 2.0.0
 * @category transformation
 */
export const FromBabbageTransactionOutputCDDLSchema = /*#__PURE__*/Schema.transformOrFail(BabbageTransactionOutputCDDL, /*#__PURE__*/Schema.typeSchema(BabbageTransactionOutput), {
  strict: true,
  encode: toI => E.gen(function* () {
    const outputMap = new Map();
    const addressBytes = yield* encAddress(toI.address);
    const valueBytes = yield* encValue(toI.amount);
    // Prepare optional fields
    const datumOptionBytes = toI.datumOption !== undefined ? yield* encDatumOption(toI.datumOption) : undefined;
    const scriptRefBytes = toI.scriptRef !== undefined ? yield* encScriptRef(toI.scriptRef) : undefined;
    // Build result object with conditional properties
    outputMap.set(0n, addressBytes);
    outputMap.set(1n, valueBytes);
    if (datumOptionBytes !== undefined) {
      outputMap.set(2n, datumOptionBytes);
    }
    if (scriptRefBytes !== undefined) {
      outputMap.set(3n, scriptRefBytes);
    }
    return outputMap;
  }),
  decode: fromI => E.gen(function* () {
    // Assume `fromI` is a CBOR Map and read keys directly.
    const addressBytes = fromI.get(0n);
    const valueBytes = fromI.get(1n);
    const datumOptionBytes = fromI.get(2n);
    const scriptRefBytes = fromI.get(3n);
    const address = yield* decAddress(addressBytes);
    const amount = yield* decValue(valueBytes);
    const datumOption = datumOptionBytes !== undefined ? yield* decDatumOption(datumOptionBytes) : undefined;
    const scriptRef = scriptRefBytes !== undefined ? yield* decScriptRef(scriptRefBytes) : undefined;
    return new BabbageTransactionOutput({
      address,
      amount,
      datumOption,
      scriptRef
    }, {
      disableValidation: true
    });
  })
});
export const CDDLSchema = /*#__PURE__*/Schema.Union(ShelleyTransactionOutputCDDL, BabbageTransactionOutputCDDL);
/**
 * CDDL schema for transaction outputs
 *
 * @since 2.0.0
 * @category transformer
 */
export const FromCDDL = /*#__PURE__*/Schema.Union(FromShelleyTransactionOutputCDDLSchema, FromBabbageTransactionOutputCDDLSchema);
/**
 * CBOR bytes transformation schema for TransactionOutput.
 *
 * @since 2.0.0
 * @category transformer
 */
export const FromCBORBytes = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.compose(CBOR.FromBytes(options),
// Uint8Array → CBOR
FromCDDL // CBOR → TransactionOutput
).annotations({
  identifier: "TransactionOutput.FromCBORBytes",
  title: "TransactionOutput from CBOR Bytes",
  description: "Transforms CBOR bytes (Uint8Array) to TransactionOutput"
});
/**
 * CBOR hex transformation schema for TransactionOutput.
 *
 * @since 2.0.0
 * @category transformer
 */
export const FromCBORHex = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.compose(Schema.Uint8ArrayFromHex,
// string → Uint8Array
FromCBORBytes(options) // Uint8Array → TransactionOutput
).annotations({
  identifier: "TransactionOutput.FromCBORHex",
  title: "TransactionOutput from CBOR Hex",
  description: "Transforms CBOR hex string to TransactionOutput"
});
/**
 * @since 2.0.0
 * @category arbitrary
 */
export const arbitrary = /*#__PURE__*/FastCheck.oneof(
/*#__PURE__*/
// Shelley TransactionOutput
FastCheck.record({
  address: FastCheck.oneof(BaseAddress.arbitrary, EnterpriseAddress.arbitrary),
  amount: Value.arbitrary,
  datumHash: FastCheck.option(DatumHash.arbitrary, {
    nil: undefined
  })
}).map(props => new ShelleyTransactionOutput(props)),
/*#__PURE__*/
// Babbage TransactionOutput
FastCheck.record({
  address: FastCheck.oneof(BaseAddress.arbitrary, EnterpriseAddress.arbitrary),
  amount: Value.arbitrary,
  datumOption: FastCheck.option(DatumOption.arbitrary, {
    nil: undefined
  }),
  scriptRef: FastCheck.option(ScriptRef.arbitrary, {
    nil: undefined
  })
}).map(props => new BabbageTransactionOutput(props)));
/**
 * Convert TransactionOutput to CBOR bytes.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toCBORBytes = (data, options = CBOR.CML_DEFAULT_OPTIONS) => Schema.encodeSync(FromCBORBytes(options))(data);
/**
 * Convert TransactionOutput to CBOR hex.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toCBORHex = (data, options = CBOR.CML_DEFAULT_OPTIONS) => Schema.encodeSync(FromCBORHex(options))(data);
/**
 * Parse TransactionOutput from CBOR bytes.
 *
 * @since 2.0.0
 * @category decoding
 */
export const fromCBORBytes = (bytes, options = CBOR.CML_DEFAULT_OPTIONS) => Schema.decodeSync(FromCBORBytes(options))(bytes);
/**
 * Parse TransactionOutput from CBOR hex.
 *
 * @since 2.0.0
 * @category decoding
 */
export const fromCBORHex = (hex, options = CBOR.CML_DEFAULT_OPTIONS) => Schema.decodeSync(FromCBORHex(options))(hex);
