// @evolution-sdk/evolution/dist/TransactionMetadatum.js
import { FastCheck, Schema } from "effect";
import * as CBOR from "./CBOR.js";
import * as Numeric from "./Numeric.js";
/**
 * Schema for TransactionMetadatum map type
 *
 * @category schemas
 * @since 2.0.0
 */
export const MapSchema = /*#__PURE__*/Schema.Map({
  key: Schema.suspend(() => TransactionMetadatumSchema).annotations({
    identifier: "TransactionMetadatum.Map.Key",
    title: "Map Key",
    description: "The key of the metadata map, must be a TransactionMetadatum type"
  }),
  value: Schema.suspend(() => TransactionMetadatumSchema).annotations({
    identifier: "TransactionMetadatum.Map.Value",
    title: "Map Value",
    description: "The value of the metadata map, must be a TransactionMetadatum type"
  })
}).annotations({
  identifier: "TransactionMetadatum.Map",
  title: "Metadata Map",
  description: "A map of TransactionMetadatum key-value pairs"
});
/**
 * Schema for TransactionMetadatum list type
 *
 * @category schemas
 * @since 2.0.0
 */
export const ListSchema = /*#__PURE__*/Schema.Array(Schema.suspend(() => TransactionMetadatumSchema)).annotations({
  identifier: "TransactionMetadatum.List",
  title: "Metadata List",
  description: "An array of TransactionMetadatum values"
});
/**
 * Schema for TransactionMetadatum string type
 *
 * @category schemas
 * @since 2.0.0
 */
export const TextSchema = /*#__PURE__*/Schema.String.annotations({
  identifier: "TransactionMetadatum.Text",
  title: "Metadata Text",
  description: "A text string value in transaction metadata"
});
/**
 * Schema for TransactionMetadatum integer type
 *
 * @category schemas
 * @since 2.0.0
 */
export const IntSchema = /*#__PURE__*/Numeric.Int64.annotations({
  identifier: "TransactionMetadatum.Int",
  title: "Metadata Integer",
  description: "An integer value in transaction metadata (64-bit signed)"
});
/**
 * Schema for TransactionMetadatum bytes type
 *
 * @category schemas
 * @since 2.0.0
 */
export const BytesSchema = /*#__PURE__*/Schema.Uint8ArrayFromHex.annotations({
  identifier: "TransactionMetadatum.Bytes",
  title: "Metadata Bytes",
  description: "A byte array value in transaction metadata"
});
/**
 * Union schema for all types of transaction metadata.
 *
 * @since 2.0.0
 * @category schemas
 */
export const TransactionMetadatumSchema = /*#__PURE__*/Schema.Union(
// Map: ReadonlyArray<[E, E]> <-> Map<T, T>
MapSchema,
// List: ReadonlyArray<E> <-> ReadonlyArray<T>
ListSchema,
// Int: string <-> bigint
IntSchema,
// Bytes: hex string <-> Uint8Array
BytesSchema,
// Text: string <-> string
TextSchema).annotations({
  identifier: "TransactionMetadatum",
  description: "Transaction metadata supporting text, integers, bytes, arrays, and maps"
});
// ============================================================================
// CBOR Functions
// ============================================================================
/**
 * Schema transformer for TransactionMetadatum from CBOR bytes.
 *
 * Uses Schema.typeSchema(TransactionMetadatumSchema) because CBOR.FromBytes
 * returns runtime types (bigint, Uint8Array, Map), not encoded types.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromCBORBytes = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.compose(CBOR.FromBytes(options), Schema.typeSchema(TransactionMetadatumSchema)).annotations({
  identifier: "TransactionMetadatum.FromCBORBytes",
  description: "Transforms CBOR bytes to TransactionMetadatum"
});
/**
 * Schema transformer for TransactionMetadatum from CBOR hex string.
 *
 * @since 2.0.0
 * @category schemas
 */
export const FromCBORHex = (options = CBOR.CML_DEFAULT_OPTIONS) => Schema.compose(Schema.Uint8ArrayFromHex, FromCBORBytes(options)).annotations({
  identifier: "TransactionMetadatum.FromCBORHex",
  description: "Transforms CBOR hex string to TransactionMetadatum"
});
// ============================================================================
// Utility Functions
// ============================================================================
/**
 * Schema-derived structural equality for TransactionMetadatum values.
 * Handles maps, lists, ints, bytes, and text via the
 * recursive TransactionMetadatumSchema definition — no hand-rolled comparison needed.
 *
 * @since 2.0.0
 * @category equality
 */
export const equals = /*#__PURE__*/Schema.equivalence(TransactionMetadatumSchema);
/**
 * FastCheck arbitrary for generating random TransactionMetadatum instances.
 *
 * @since 2.0.0
 * @category testing
 */
const I64_MIN = -(1n << 63n);
const I64_MAX = (1n << 63n) - 1n;
const int64Arbitrary = /*#__PURE__*/FastCheck.bigInt({
  min: I64_MIN,
  max: I64_MAX
});
export const arbitrary = /*#__PURE__*/FastCheck.oneof(/*#__PURE__*/FastCheck.string(), int64Arbitrary, /*#__PURE__*/FastCheck.uint8Array({
  minLength: 1,
  maxLength: 64
}), /*#__PURE__*/FastCheck.array(/*#__PURE__*/FastCheck.oneof(/*#__PURE__*/FastCheck.string(), int64Arbitrary), {
  maxLength: 5
}), /*#__PURE__*/FastCheck.uniqueArray(FastCheck.tuple(FastCheck.string(), int64Arbitrary), {
  maxLength: 5,
  selector: ([key]) => key // Ensure unique keys
}).map(entries => new globalThis.Map(entries)));
// ============================================================================
// Parsing Functions
// ============================================================================
/**
 * Parse a TransactionMetadatum from CBOR bytes.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromCBORBytes = (bytes, options = CBOR.CML_DEFAULT_OPTIONS) => Schema.decodeSync(FromCBORBytes(options))(bytes);
/**
 * Parse a TransactionMetadatum from CBOR hex string.
 *
 * @since 2.0.0
 * @category parsing
 */
export const fromCBORHex = (hex, options = CBOR.CML_DEFAULT_OPTIONS) => Schema.decodeSync(FromCBORHex(options))(hex);
// ============================================================================
// Encoding Functions
// ============================================================================
/**
 * Convert a TransactionMetadatum to CBOR bytes.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toCBORBytes = (data, options = CBOR.CML_DEFAULT_OPTIONS) => Schema.encodeSync(FromCBORBytes(options))(data);
/**
 * Convert a TransactionMetadatum to CBOR hex string.
 *
 * @since 2.0.0
 * @category encoding
 */
export const toCBORHex = (data, options = CBOR.CML_DEFAULT_OPTIONS) => Schema.encodeSync(FromCBORHex(options))(data);
// ============================================================================
// Factory Functions
// ============================================================================
/**
 * Create a text TransactionMetadatum from a string value.
 *
 * @since 2.0.0
 * @category constructors
 */
export const text = value => value;
/**
 * Create an integer TransactionMetadatum from a bigint value.
 *
 * @since 2.0.0
 * @category constructors
 */
export const int = value => value;
/**
 * Create a bytes TransactionMetadatum from a Uint8Array value.
 *
 * @since 2.0.0
 * @category constructors
 */
export const bytes = value => value;
/**
 * Create an array TransactionMetadatum from an array of TransactionMetadatum values.
 *
 * @since 2.0.0
 * @category constructors
 */
export const array = value => value;
/**
 * Create a map TransactionMetadatum from a Map of TransactionMetadatum key-value pairs.
 *
 * @since 2.0.0
 * @category constructors
 */
export const map = value => value;
/**
 * Create a map TransactionMetadatum from an array of key-value pair entries.
 *
 * @since 2.0.0
 * @category constructors
 */
export const fromEntries = entries => new globalThis.Map(entries);
