// @evolution-sdk/evolution/dist/message-signing/Label.js
/**
 * COSE Label types and constructors.
 *
 * @since 2.0.0
 * @category Message Signing
 */
import { Equal, Hash, Inspectable, Schema } from "effect";
// ============================================================================
// Enums
// ============================================================================
/**
 * COSE Algorithm Identifiers (RFC 8152).
 *
 * @since 2.0.0
 * @category Enums
 */
export var AlgorithmId;
(function (AlgorithmId) {
  AlgorithmId[AlgorithmId["EdDSA"] = -8] = "EdDSA";
})(AlgorithmId || (AlgorithmId = {}));
/**
 * COSE Key Type values (RFC 8152).
 *
 * @since 2.0.0
 * @category Enums
 */
export var KeyType;
(function (KeyType) {
  KeyType[KeyType["OKP"] = 1] = "OKP";
  KeyType[KeyType["EC2"] = 2] = "EC2";
  KeyType[KeyType["RSA"] = 3] = "RSA";
  KeyType[KeyType["Symmetric"] = 4] = "Symmetric";
  KeyType[KeyType["HSS_LMS"] = 5] = "HSS_LMS";
  KeyType[KeyType["WalnutDSA"] = 6] = "WalnutDSA";
})(KeyType || (KeyType = {}));
/**
 * COSE Curve Type values (RFC 8152).
 *
 * @since 2.0.0
 * @category Enums
 */
export var CurveType;
(function (CurveType) {
  CurveType[CurveType["P256"] = 1] = "P256";
  CurveType[CurveType["P384"] = 2] = "P384";
  CurveType[CurveType["P521"] = 3] = "P521";
  CurveType[CurveType["X25519"] = 4] = "X25519";
  CurveType[CurveType["X448"] = 5] = "X448";
  CurveType[CurveType["Ed25519"] = 6] = "Ed25519";
  CurveType[CurveType["Ed448"] = 7] = "Ed448";
  CurveType[CurveType["Secp256k1"] = 8] = "Secp256k1";
})(CurveType || (CurveType = {}));
/**
 * COSE Key Operations (RFC 8152).
 *
 * @since 2.0.0
 * @category Enums
 */
export var KeyOperation;
(function (KeyOperation) {
  KeyOperation[KeyOperation["Sign"] = 0] = "Sign";
  KeyOperation[KeyOperation["Verify"] = 1] = "Verify";
  KeyOperation[KeyOperation["Encrypt"] = 2] = "Encrypt";
  KeyOperation[KeyOperation["Decrypt"] = 3] = "Decrypt";
  KeyOperation[KeyOperation["WrapKey"] = 4] = "WrapKey";
  KeyOperation[KeyOperation["UnwrapKey"] = 5] = "UnwrapKey";
  KeyOperation[KeyOperation["DeriveKey"] = 6] = "DeriveKey";
  KeyOperation[KeyOperation["DeriveBits"] = 7] = "DeriveBits";
  KeyOperation[KeyOperation["MacCreate"] = 8] = "MacCreate";
  KeyOperation[KeyOperation["MacVerify"] = 9] = "MacVerify";
})(KeyOperation || (KeyOperation = {}));
/**
 * Signature context for Sig_structure (RFC 8152).
 *
 * @since 2.0.0
 * @category Enums
 */
export var SigContext;
(function (SigContext) {
  SigContext["Signature"] = "Signature";
  SigContext["Signature1"] = "Signature1";
  SigContext["CounterSignature"] = "CounterSignature";
})(SigContext || (SigContext = {}));
/**
 * Label kind discriminator.
 *
 * @since 2.0.0
 * @category Enums
 */
export var LabelKind;
(function (LabelKind) {
  LabelKind[LabelKind["Int"] = 0] = "Int";
  LabelKind[LabelKind["Text"] = 1] = "Text";
})(LabelKind || (LabelKind = {}));
// ============================================================================
// Label
// ============================================================================
/**
 * COSE header label - can be an integer or text string (RFC 8152).
 *
 * @since 2.0.0
 * @category Model
 */
export class Label extends /*#__PURE__*/Schema.Class("Label")({
  kind: /*#__PURE__*/Schema.Enums(LabelKind),
  value: /*#__PURE__*/Schema.Union(Schema.BigIntFromSelf, Schema.String)
}) {
  toJSON() {
    return {
      _tag: "Label",
      kind: this.kind === LabelKind.Int ? "Int" : "Text",
      value: this.kind === LabelKind.Int ? this.value.toString() : this.value
    };
  }
  toString() {
    return Inspectable.format(this.toJSON());
  }
  [Inspectable.NodeInspectSymbol]() {
    return this.toJSON();
  }
  [Equal.symbol](that) {
    return that instanceof Label && this.kind === that.kind && (this.kind === LabelKind.Int && that.kind === LabelKind.Int ? this.value === that.value : this.value === that.value);
  }
  [Hash.symbol]() {
    return Hash.combine(Hash.number(this.kind))(typeof this.value === "bigint" ? Hash.number(Number(this.value)) : Hash.string(this.value));
  }
  /**
   * Get the integer value (throws if label is text).
   *
   * @since 2.0.0
   * @category Accessors
   */
  asInt() {
    if (this.kind !== LabelKind.Int) {
      throw new Error("Label is not an integer");
    }
    return this.value;
  }
  /**
   * Get the text value (throws if label is integer).
   *
   * @since 2.0.0
   * @category Accessors
   */
  asText() {
    if (this.kind !== LabelKind.Text) {
      throw new Error("Label is not text");
    }
    return this.value;
  }
}
// ============================================================================
// Constructors
// ============================================================================
/**
 * Create a Label from an integer.
 *
 * @since 2.0.0
 * @category Constructors
 */
export const labelFromInt = value => new Label({
  kind: LabelKind.Int,
  value
}, {
  disableValidation: true
});
/**
 * Create a Label from a text string.
 *
 * @since 2.0.0
 * @category Constructors
 */
export const labelFromText = value => new Label({
  kind: LabelKind.Text,
  value
}, {
  disableValidation: true
});
/**
 * Create a Label from AlgorithmId.
 *
 * @since 2.0.0
 * @category Constructors
 */
export const labelFromAlgorithmId = alg => labelFromInt(BigInt(alg));
/**
 * Create a Label from KeyType.
 *
 * @since 2.0.0
 * @category Constructors
 */
export const labelFromKeyType = kty => labelFromInt(BigInt(kty));
/**
 * Create a Label from CurveType.
 *
 * @since 2.0.0
 * @category Constructors
 */
export const labelFromCurveType = crv => labelFromInt(BigInt(crv));
