// @evolution-sdk/evolution/dist/message-signing/Utils.js
/**
 * Utility functions for COSE message signing.
 *
 * @since 2.0.0
 * @category Message Signing
 */
import { FastCheck } from "effect";
import * as Bytes from "../Bytes.js";
// ============================================================================
// Payload Conversion
// ============================================================================
/**
 * Convert text to Payload (UTF-8 encoding).
 *
 * @since 2.0.0
 * @category Utilities
 */
export const fromText = text => new TextEncoder().encode(text);
/**
 * Convert Payload to text (UTF-8 decoding).
 *
 * @since 2.0.0
 * @category Utilities
 */
export const toText = payload => new TextDecoder().decode(payload);
/**
 * Convert hex string to Payload.
 *
 * @since 2.0.0
 * @category Utilities
 */
export const fromHex = hex => Bytes.fromHex(hex);
/**
 * Convert Payload to hex string.
 *
 * @since 2.0.0
 * @category Utilities
 */
export const toHex = payload => Bytes.toHex(payload);
// ============================================================================
// Testing Support
// ============================================================================
/**
 * FastCheck arbitrary for generating random Payload instances.
 *
 * @since 2.0.0
 * @category Testing
 */
export const arbitraryPayload = /*#__PURE__*/FastCheck.uint8Array({
  minLength: 0,
  maxLength: 256
});
// ============================================================================
// Internal Utilities
// ============================================================================
/**
 * FNV-1a 32-bit hash algorithm.
 * Used for checksums in user-facing encoding.
 *
 * @internal
 */
export function fnv32a(bytes) {
  const FNV_PRIME = 0x01000193;
  let hash = 0x811c9dc5; // FNV offset basis
  for (let i = 0; i < bytes.length; i++) {
    hash ^= bytes[i];
    hash = Math.imul(hash, FNV_PRIME);
  }
  return hash >>> 0; // Convert to unsigned 32-bit
}
