// @effect/cluster/dist/esm/RunnerHealth.js
/**
 * @since 1.0.0
 */
import * as Context from "effect/Context";
import * as Effect from "effect/Effect";
import * as Layer from "effect/Layer";
import * as RcMap from "effect/RcMap";
import * as MessageStorage from "./MessageStorage.js";
import * as Runners from "./Runners.js";
/**
 * Represents the service used to check if a Runner is healthy.
 *
 * If a Runner is responsive, shards will not be re-assigned because the Runner may
 * still be processing messages. If a Runner is not responsive, then its
 * associated shards can and will be re-assigned to a different Runner.
 *
 * @since 1.0.0
 * @category models
 */
export class RunnerHealth extends /*#__PURE__*/Context.Tag("@effect/cluster/RunnerHealth")() {}
/**
 * @since 1.0.0
 * @category Constructors
 */
export const make = /*#__PURE__*/Effect.fnUntraced(function* (options) {
  const connections = yield* RcMap.make({
    lookup: _address => Effect.void
  });
  const onConnection = address => RcMap.get(connections, address);
  const isAlive = Effect.fnUntraced(function* (address) {
    if (yield* RcMap.has(connections, address)) {
      return true;
    }
    return yield* options.isAlive(address);
  });
  return RunnerHealth.of({
    onConnection,
    isAlive
  });
});
/**
 * A layer which will **always** consider a Runner healthy.
 *
 * This is useful for testing.
 *
 * @since 1.0.0
 * @category layers
 */
export const layerNoop = /*#__PURE__*/Layer.scoped(RunnerHealth, /*#__PURE__*/make({
  isAlive: () => Effect.succeed(true)
}));
/**
 * @since 1.0.0
 * @category Constructors
 */
export const makePing = /*#__PURE__*/Effect.gen(function* () {
  const runners = yield* Runners.Runners;
  function isAlive(address) {
    return runners.ping(address).pipe(Effect.timeout(3000), Effect.retry({
      times: 3
    }), Effect.isSuccess);
  }
  return yield* make({
    isAlive
  });
});
/**
 * A layer which will ping a Runner directly to check if it is healthy.
 *
 * @since 1.0.0
 * @category layers
 */
export const layer = /*#__PURE__*/Layer.scoped(RunnerHealth, makePing);
/**
 * A layer which will ping a Runner directly to check if it is healthy.
 *
 * @since 1.0.0
 * @category layers
 */
export const layerRpc = /*#__PURE__*/layer.pipe(/*#__PURE__*/Layer.provide(Runners.layerRpc), /*#__PURE__*/Layer.provide(MessageStorage.layerNoop));
