// @effect/cluster/dist/esm/ShardingConfig.js
/**
 * @since 1.0.0
 */
import * as Config from "effect/Config";
import * as ConfigProvider from "effect/ConfigProvider";
import * as Context from "effect/Context";
import * as Duration from "effect/Duration";
import * as Effect from "effect/Effect";
import * as Layer from "effect/Layer";
import * as Option from "effect/Option";
import { RunnerAddress } from "./RunnerAddress.js";
/**
 * Represents the configuration for the `Sharding` service on a given runner.
 *
 * @since 1.0.0
 * @category models
 */
export class ShardingConfig extends /*#__PURE__*/Context.Tag("@effect/cluster/ShardingConfig")() {}
const defaultRunnerAddress = /*#__PURE__*/RunnerAddress.make({
  host: "localhost",
  port: 34431
});
/**
 * @since 1.0.0
 * @category defaults
 */
export const defaults = {
  runnerAddress: /*#__PURE__*/Option.some(defaultRunnerAddress),
  runnerListenAddress: /*#__PURE__*/Option.none(),
  serverVersion: 1,
  shardsPerGroup: 300,
  shardManagerAddress: /*#__PURE__*/RunnerAddress.make({
    host: "localhost",
    port: 8080
  }),
  shardManagerUnavailableTimeout: /*#__PURE__*/Duration.minutes(10),
  shardGroups: ["default"],
  entityMailboxCapacity: 4096,
  entityMaxIdleTime: /*#__PURE__*/Duration.minutes(1),
  entityTerminationTimeout: /*#__PURE__*/Duration.seconds(15),
  entityMessagePollInterval: /*#__PURE__*/Duration.seconds(10),
  entityReplyPollInterval: /*#__PURE__*/Duration.millis(200),
  sendRetryInterval: /*#__PURE__*/Duration.millis(100),
  refreshAssignmentsInterval: /*#__PURE__*/Duration.minutes(5),
  simulateRemoteSerialization: true
};
/**
 * @since 1.0.0
 * @category Layers
 */
export const layer = options => Layer.succeed(ShardingConfig, {
  ...defaults,
  ...options
});
/**
 * @since 1.0.0
 * @category defaults
 */
export const layerDefaults = /*#__PURE__*/layer();
/**
 * @since 1.0.0
 * @category Config
 */
export const config = /*#__PURE__*/Config.all({
  runnerAddress: /*#__PURE__*/Config.all({
    host: Config.string("host").pipe(Config.withDefault(defaultRunnerAddress.host), Config.withDescription("The hostname or IP address of the runner.")),
    port: Config.integer("port").pipe(Config.withDefault(defaultRunnerAddress.port), Config.withDescription("The port used for inter-runner communication."))
  }).pipe(/*#__PURE__*/Config.map(options => RunnerAddress.make(options)), Config.option),
  runnerListenAddress: /*#__PURE__*/Config.all({
    host: Config.string("listenHost").pipe(Config.withDescription("The host to listen on.")),
    port: Config.integer("listenPort").pipe(Config.withDefault(defaultRunnerAddress.port), Config.withDescription("The port to listen on."))
  }).pipe(/*#__PURE__*/Config.map(options => RunnerAddress.make(options)), Config.option),
  serverVersion: /*#__PURE__*/Config.integer("serverVersion").pipe(/*#__PURE__*/Config.withDefault(defaults.serverVersion), /*#__PURE__*/Config.withDescription("The version of the current runner.")),
  shardGroups: /*#__PURE__*/Config.array(Config.string("shardGroups")).pipe(/*#__PURE__*/Config.withDefault(["default"]), /*#__PURE__*/Config.withDescription("The shard groups that are assigned to this runner.")),
  shardsPerGroup: /*#__PURE__*/Config.integer("shardsPerGroup").pipe(/*#__PURE__*/Config.withDefault(defaults.shardsPerGroup), /*#__PURE__*/Config.withDescription("The number of shards to allocate per shard group.")),
  shardManagerAddress: /*#__PURE__*/Config.all({
    host: Config.string("shardManagerHost").pipe(Config.withDefault(defaults.shardManagerAddress.host), Config.withDescription("The host of the shard manager.")),
    port: Config.integer("shardManagerPort").pipe(Config.withDefault(defaults.shardManagerAddress.port), Config.withDescription("The port of the shard manager."))
  }).pipe(/*#__PURE__*/Config.map(options => RunnerAddress.make(options))),
  shardManagerUnavailableTimeout: /*#__PURE__*/Config.duration("shardManagerUnavailableTimeout").pipe(/*#__PURE__*/Config.withDefault(defaults.shardManagerUnavailableTimeout), /*#__PURE__*/Config.withDescription("If the shard is unavilable for this duration, all the shard assignments will be reset.")),
  entityMailboxCapacity: /*#__PURE__*/Config.integer("entityMailboxCapacity").pipe(/*#__PURE__*/Config.withDefault(defaults.entityMailboxCapacity), /*#__PURE__*/Config.withDescription("The default capacity of the mailbox for entities.")),
  entityMaxIdleTime: /*#__PURE__*/Config.duration("entityMaxIdleTime").pipe(/*#__PURE__*/Config.withDefault(defaults.entityMaxIdleTime), /*#__PURE__*/Config.withDescription("The maximum duration of inactivity (i.e. without receiving a message) after which an entity will be interrupted.")),
  entityTerminationTimeout: /*#__PURE__*/Config.duration("entityTerminationTimeout").pipe(/*#__PURE__*/Config.withDefault(defaults.entityTerminationTimeout), /*#__PURE__*/Config.withDescription("The maximum duration of time to wait for an entity to terminate.")),
  entityMessagePollInterval: /*#__PURE__*/Config.duration("entityMessagePollInterval").pipe(/*#__PURE__*/Config.withDefault(defaults.entityMessagePollInterval), /*#__PURE__*/Config.withDescription("The interval at which to poll for unprocessed messages from storage.")),
  entityReplyPollInterval: /*#__PURE__*/Config.duration("entityReplyPollInterval").pipe(/*#__PURE__*/Config.withDefault(defaults.entityReplyPollInterval), /*#__PURE__*/Config.withDescription("The interval at which to poll for client replies from storage.")),
  sendRetryInterval: /*#__PURE__*/Config.duration("sendRetryInterval").pipe(/*#__PURE__*/Config.withDefault(defaults.sendRetryInterval), /*#__PURE__*/Config.withDescription("The interval to retry a send if EntityNotManagedByRunner is returned.")),
  refreshAssignmentsInterval: /*#__PURE__*/Config.duration("refreshAssignmentsInterval").pipe(/*#__PURE__*/Config.withDefault(defaults.refreshAssignmentsInterval), /*#__PURE__*/Config.withDescription("The interval at which to refresh shard assignments.")),
  simulateRemoteSerialization: /*#__PURE__*/Config.boolean("simulateRemoteSerialization").pipe(/*#__PURE__*/Config.withDefault(defaults.simulateRemoteSerialization), /*#__PURE__*/Config.withDescription("Simulate serialization and deserialization to remote runners for local entities."))
});
/**
 * @since 1.0.0
 * @category Config
 */
export const configFromEnv = /*#__PURE__*/config.pipe(/*#__PURE__*/Effect.withConfigProvider(/*#__PURE__*/ConfigProvider.fromEnv().pipe(ConfigProvider.constantCase)));
/**
 * @since 1.0.0
 * @category Layers
 */
export const layerFromEnv = options => Layer.effect(ShardingConfig, options ? Effect.map(configFromEnv, config => ({
  ...config,
  ...options
})) : configFromEnv);
