// @evolution-sdk/evolution/dist/sdk/builders/RedeemerBuilder.js
/**
 * Deferred redeemer construction for dynamic index resolution.
 *
 * Cardano redeemers must reference inputs/policies by their sorted index in the
 * final transaction. But coin selection may add new inputs, changing all indices.
 * This creates a chicken-and-egg problem: you need indices to build redeemers,
 * but indices aren't known until the transaction is finalized.
 *
 * The solution is three modes:
 * - Static: When you know the redeemer value upfront (no index needed)
 * - Self: When redeemer depends on THIS input's final index
 * - Batch: When redeemer depends on MULTIPLE inputs' final indices
 *
 * @module RedeemerBuilder
 * @since 2.0.0
 */
/** @since 2.0.0 @category guards */
export const isSelfFn = arg => typeof arg === "function";
/** @since 2.0.0 @category guards */
export const isBatchBuilder = arg => typeof arg === "object" && arg !== null && "all" in arg && typeof arg.all === "function";
/** @since 2.0.0 @category guards */
export const isStaticData = arg => !isSelfFn(arg) && !isBatchBuilder(arg);
/** @internal */
export const toDeferredRedeemer = arg => {
  if (isSelfFn(arg)) {
    return {
      _tag: "self",
      fn: arg
    };
  }
  if (isBatchBuilder(arg)) {
    return {
      _tag: "batch",
      fn: arg.all,
      inputs: arg.inputs
    };
  }
  return {
    _tag: "static",
    data: arg
  };
};
