// @evolution-sdk/evolution/dist/sdk/builders/TransactionBuilder.js
/**
 * Transaction builder storing a sequence of deferred operations that assemble and balance a transaction.
 *
 * @module TransactionBuilder
 * @since 2.0.0
 *
 * ## Execution Model
 *
 * The builder pattern:
 * - **Immutable configuration** at construction (protocol params, change address, available UTxOs)
 * - **ProgramSteps array** accumulates deferred effects via chainable API methods
 * - **Fresh state per build()** — each execution creates new Ref instances, runs all programs sequentially
 * - **Deferred composition** — no I/O or state updates occur until build() is invoked
 *
 * Key invariant: calling `build()` twice with the same builder instance produces two independent results
 * with no cross-contamination because fresh state (Refs) is created each time.
 *
 * ## Coin Selection
 *
 * Automatic coin selection selects UTxOs from `availableUtxos` to satisfy transaction outputs and fees.
 * The `collectFrom()` method allows manual input selection; automatic selection excludes these to prevent
 * double-spending. UTxOs can come from any source (wallet, DeFi protocols, other participants, etc.).
 *
 * @since 2.0.0
 */
// Effect-TS imports
import { Context, Data, Effect, Layer, Logger, LogLevel, Ref } from "effect";
import * as CoreAssets from "../../Assets/index.js";
import * as Time from "../../Time/index.js";
import * as Transaction from "../../Transaction.js";
import { runEffectPromise } from "../../utils/effect-runtime.js";
import { createAddSignerProgram } from "./operations/AddSigner.js";
import { attachScriptToState } from "./operations/Attach.js";
import { createAttachMetadataProgram } from "./operations/AttachMetadata.js";
import { createCollectFromProgram } from "./operations/Collect.js";
import { createAuthCommitteeHotProgram, createDeregisterDRepProgram, createRegisterDRepProgram, createResignCommitteeColdProgram, createUpdateDRepProgram } from "./operations/Governance.js";
import { createMintAssetsProgram } from "./operations/Mint.js";
import { createPayToAddressProgram } from "./operations/Pay.js";
import { createRegisterPoolProgram, createRetirePoolProgram } from "./operations/Pool.js";
import { createProposeProgram } from "./operations/Propose.js";
import { createReadFromProgram } from "./operations/ReadFrom.js";
import { createSendAllProgram } from "./operations/SendAll.js";
import { createDelegateToDRepProgram, createDelegateToPoolAndDRepProgram, createDelegateToPoolProgram, createDelegateToProgram, createDeregisterStakeProgram, createRegisterAndDelegateToProgram, createRegisterStakeProgram, createWithdrawProgram } from "./operations/Stake.js";
import { createSetValidityProgram } from "./operations/Validity.js";
import { createVoteProgram } from "./operations/Vote.js";
import { executeBalance } from "./phases/Balance.js";
import { executeChangeCreation } from "./phases/ChangeCreation.js";
import { executeCollateral } from "./phases/Collateral.js";
import { executeEvaluation } from "./phases/Evaluation.js";
import { executeFallback } from "./phases/Fallback.js";
import { executeFeeCalculation } from "./phases/FeeCalculation.js";
import { executeSelection } from "./phases/Selection.js";
import { makeSignBuilder } from "./SignBuilderImpl.js";
import { makeTransactionResult } from "./TransactionResult.js";
import { assembleTransaction, buildFakeWitnessSet, buildTransactionInputs, calculateTransactionSize } from "./TxBuilderImpl.js";
/**
 * Error type for failures occurring during transaction builder operations.
 *
 * @since 2.0.0
 * @category errors
 */
export class TransactionBuilderError extends /*#__PURE__*/Data.TaggedError("TransactionBuilderError") {}
export class PhaseContextTag extends /*#__PURE__*/Context.Tag("PhaseContextTag")() {}
// Initial state for transaction builder
const initialTxBuilderState = {
  selectedUtxos: [],
  outputs: [],
  scripts: /*#__PURE__*/new Map(),
  totalOutputAssets: CoreAssets.zero,
  totalInputAssets: CoreAssets.zero,
  redeemers: /*#__PURE__*/new Map(),
  deferredRedeemers: /*#__PURE__*/new Map(),
  referenceInputs: [],
  certificates: [],
  withdrawals: /*#__PURE__*/new Map(),
  poolDeposits: /*#__PURE__*/new Map(),
  requiredSigners: [],
  auxiliaryData: undefined
};
/**
 * Resolve protocol parameters from options, provider, or fail.
 * Priority: BuildOptions override > provider.getProtocolParameters() > error
 */
const resolveProtocolParameters = (config, options) => {
  if (options?.protocolParameters !== undefined) {
    return Effect.succeed(options.protocolParameters);
  }
  if (config.provider) {
    return Effect.map(config.provider.Effect.getProtocolParameters(), params => ({
      minFeeCoefficient: BigInt(params.minFeeA),
      minFeeConstant: BigInt(params.minFeeB),
      coinsPerUtxoByte: params.coinsPerUtxoByte,
      maxTxSize: params.maxTxSize,
      priceMem: params.priceMem,
      priceStep: params.priceStep,
      minFeeRefScriptCostPerByte: params.minFeeRefScriptCostPerByte
    }));
  }
  return Effect.fail(new TransactionBuilderError({
    message: "No protocol parameters provided. Either provide protocolParameters in BuildOptions or provider in config.",
    cause: null
  }));
};
/**
 * Resolve change address from options, wallet, or fail.
 * Priority: BuildOptions override > wallet.address() > error
 */
const resolveChangeAddress = (config, options) => {
  if (options?.changeAddress) {
    return Effect.succeed(options.changeAddress);
  }
  if (config.wallet) {
    return config.wallet.Effect.address();
  }
  return Effect.fail(new TransactionBuilderError({
    message: "No change address provided. Either provide wallet in config or changeAddress in build options.",
    cause: null
  }));
};
/**
 * Resolve available UTxOs from options, provider+wallet, or fail.
 * Priority: BuildOptions override > provider.getUtxos(wallet.address) > error
 */
const resolveAvailableUtxos = (config, options) => {
  if (options?.availableUtxos) {
    return Effect.succeed(options.availableUtxos);
  }
  if (config.wallet && config.provider) {
    return Effect.flatMap(config.wallet.Effect.address(), addr => config.provider.Effect.getUtxos(addr));
  }
  return Effect.fail(new TransactionBuilderError({
    message: "No available UTxOs provided. Either provide wallet+provider in config or availableUtxos in build options.",
    cause: null
  }));
};
/**
 * Parse Ogmios/provider error response into raw ScriptFailure array.
 * Returns failures without labels - enrichment happens in Evaluation phase.
 */
const parseProviderError = error => {
  const failures = [];
  // Navigate through error chain to find the response body
  const findErrorData = e => {
    if (!e || typeof e !== "object") return undefined;
    const obj = e;
    // Direct data property (from ProviderError cause chain)
    if (obj.cause && typeof obj.cause === "object") {
      const cause = obj.cause;
      // ResponseError with response.body
      if (cause.response && typeof cause.response === "object") {
        const resp = cause.response;
        if (resp.body && typeof resp.body === "object") {
          const body = resp.body;
          if (body.error && typeof body.error === "object") {
            const err = body.error;
            if (Array.isArray(err.data)) {
              return err.data;
            }
          }
        }
      }
      // Try description field which contains the JSON string
      if (typeof cause.description === "string") {
        try {
          const match = cause.description.match(/\{.*\}/s);
          if (match) {
            const parsed = JSON.parse(match[0]);
            if (parsed.error?.data && Array.isArray(parsed.error.data)) {
              return parsed.error.data;
            }
          }
        } catch {
          // JSON parse failed, continue looking
        }
      }
      // Recurse into cause
      return findErrorData(cause);
    }
    return undefined;
  };
  const errorData = findErrorData(error);
  if (!errorData) {
    return failures;
  }
  // Process each validator error (raw, without labels)
  for (const validatorError of errorData) {
    const {
      error: err,
      validator
    } = validatorError;
    const {
      index,
      purpose
    } = validator;
    const {
      traces,
      validationError
    } = err.data;
    failures.push({
      purpose,
      index,
      validationError,
      traces: traces ?? []
    });
  }
  return failures;
};
/**
 * Resolve evaluator from options, provider, or return undefined.
 * Priority: BuildOptions.evaluator > provider.evaluateTx (wrapped) > undefined
 *
 * When undefined is returned, the Evaluation phase will fail with an appropriate error
 * if scripts are present in the transaction.
 */
const resolveEvaluator = (config, options) => {
  // Priority 1: Explicit evaluator from BuildOptions
  if (options?.evaluator) {
    return options.evaluator;
  }
  // Priority 2: Wrap provider's evaluateTx as an Evaluator
  if (config.provider) {
    return {
      evaluate: (tx, additionalUtxos, _context) => {
        // Provider-based evaluators (Ogmios, Blockfrost) resolve UTxOs from chain.
        // By default, don't pass additionalUtxos to avoid OverlappingAdditionalUtxo errors.
        // Use passAdditionalUtxos: true for edge cases (e.g., UTxOs not yet on chain).
        const utxosToPass = options?.passAdditionalUtxos ? additionalUtxos : undefined;
        return config.provider.Effect.evaluateTx(tx, utxosToPass).pipe(Effect.mapError(providerError => {
          // Parse provider error into structured failures
          const failures = parseProviderError(providerError);
          return new EvaluationError({
            message: `Provider evaluation failed: ${providerError.message}`,
            cause: providerError,
            failures
          });
        }));
      }
    };
  }
  // No evaluator available - Evaluation phase will handle error if scripts present
  return undefined;
};
/**
 * Resolve slot configuration from BuildOptions, TxBuilderConfig, or network default.
 * Priority: BuildOptions.slotConfig > TxBuilderConfig.slotConfig > SLOT_CONFIG_NETWORK[config.network]
 *
 * Slot configuration defines the relationship between slots and Unix time,
 * required for UPLC evaluation of time-based validators and validity interval conversion.
 */
const resolveSlotConfig = (config, options) => {
  // Priority 1: Explicit slot config from BuildOptions (per-transaction override)
  if (options?.slotConfig) {
    return options.slotConfig;
  }
  // Priority 2: Slot config from TxBuilderConfig (set at client level)
  if (config.slotConfig) {
    return config.slotConfig;
  }
  // Priority 3: Network-specific slot config preset
  const network = config.network ?? "Mainnet";
  return Time.SLOT_CONFIG_NETWORK[network];
};
/**
 * Assemble final builder result based on wallet capabilities.
 * Accesses transaction data from context tags.
 */
const assembleFinalResult = (config, transaction, txWithFakeWitnesses, availableUtxos) => Effect.gen(function* () {
  const buildCtxRef = yield* PhaseContextTag;
  const buildCtx = yield* Ref.get(buildCtxRef);
  const stateRef = yield* TxContext;
  const state = yield* Ref.get(stateRef);
  const wallet = config.wallet;
  if (wallet?.type === "signing" || wallet?.type === "api") {
    return makeSignBuilder({
      transaction,
      transactionWithFakeWitnesses: txWithFakeWitnesses,
      fee: buildCtx.calculatedFee,
      utxos: state.selectedUtxos,
      referenceUtxos: state.referenceInputs,
      provider: config.provider,
      wallet,
      // Pass raw data for lazy chainResult computation
      outputs: state.outputs,
      availableUtxos
    });
  }
  return makeTransactionResult({
    transaction,
    transactionWithFakeWitnesses: txWithFakeWitnesses,
    fee: buildCtx.calculatedFee
  });
});
/**
 * Phase handler map for routing phase execution.
 * Each handler executes its specific phase logic and returns a PhaseResult indicating the next phase.
 * All phase implementations are now modularized in the phases/ directory.
 */
const phaseMap = {
  selection: executeSelection,
  changeCreation: executeChangeCreation,
  feeCalculation: executeFeeCalculation,
  balance: executeBalance,
  evaluation: executeEvaluation,
  collateral: executeCollateral,
  fallback: executeFallback
};
/**
 * Assemble and validate transaction after phase loop completes.
 */
const assembleAndValidateTransaction = /*#__PURE__*/Effect.gen(function* () {
  const buildCtxRef = yield* PhaseContextTag;
  const buildCtx = yield* Ref.get(buildCtxRef);
  const stateRef = yield* TxContext;
  yield* Effect.logDebug(`Build complete - fee: ${buildCtx.calculatedFee}`);
  // Add change outputs to the transaction outputs
  if (buildCtx.changeOutputs.length > 0) {
    yield* Ref.update(stateRef, s => ({
      ...s,
      outputs: [...s.outputs, ...buildCtx.changeOutputs]
    }));
    yield* Effect.logDebug(`Added ${buildCtx.changeOutputs.length} change output(s) to transaction`);
  }
  // Get final inputs and outputs for transaction assembly
  const finalState = yield* Ref.get(stateRef);
  const selectedUtxos = finalState.selectedUtxos;
  const allOutputs = finalState.outputs;
  yield* Effect.logDebug(`Assembling transaction: ${selectedUtxos.length} inputs, ${allOutputs.length} outputs, fee: ${buildCtx.calculatedFee}`);
  // Build transaction inputs and assemble transaction body
  const inputs = yield* buildTransactionInputs(selectedUtxos);
  const transaction = yield* assembleTransaction(inputs, allOutputs, buildCtx.calculatedFee);
  // SAFETY CHECK: Validate transaction size against protocol limit
  // Include collateral UTxOs in witness estimation - they require VKey witnesses too!
  const allUtxosForWitnesses = finalState.collateral ? [...selectedUtxos, ...finalState.collateral.inputs] : selectedUtxos;
  const fakeWitnessSet = yield* buildFakeWitnessSet(allUtxosForWitnesses);
  const txWithFakeWitnesses = new Transaction.Transaction({
    body: transaction.body,
    witnessSet: fakeWitnessSet,
    isValid: true,
    auxiliaryData: finalState.auxiliaryData ?? null
  });
  const txSizeWithWitnesses = yield* calculateTransactionSize(txWithFakeWitnesses);
  const protocolParams = yield* ProtocolParametersTag;
  yield* Effect.logDebug(`Transaction size: ${txSizeWithWitnesses} bytes ` + `(with ${fakeWitnessSet.vkeyWitnesses?.length ?? 0} fake witnesses), ` + `max=${protocolParams.maxTxSize} bytes`);
  if (txSizeWithWitnesses > protocolParams.maxTxSize) {
    return yield* Effect.fail(new TransactionBuilderError({
      message: `Transaction size (${txSizeWithWitnesses} bytes) exceeds protocol maximum (${protocolParams.maxTxSize} bytes). ` + `Consider splitting into multiple transactions.`
    }));
  }
  return {
    transaction,
    txWithFakeWitnesses
  };
});
const phaseStateMachine = /*#__PURE__*/Effect.gen(function* () {
  // Get phase context ref once (doesn't change during execution)
  const phaseContextRef = yield* PhaseContextTag;
  // Phase loop
  while (true) {
    const phaseContext = yield* Ref.get(phaseContextRef);
    // Terminal state
    if (phaseContext.phase === "complete") {
      break;
    }
    // Route to phase handler
    const phase = phaseMap[phaseContext.phase];
    if (!phase) {
      return yield* Effect.fail(new TransactionBuilderError({
        message: `Unknown phase: ${phaseContext.phase}`
      }));
    }
    const result = yield* phase();
    // Update phase
    yield* Ref.update(phaseContextRef, c => ({
      ...c,
      phase: result.next
    }));
  }
  // Assemble and validate transaction
  return yield* assembleAndValidateTransaction;
});
/**
 * Default BuildOptions for safe transaction building.
 *
 * **Safety Principles:**
 * - coinSelection: "largest-first" (deterministic, efficient)
 * - onInsufficientChange: "error" (prevents accidental fund loss)
 * - setCollateral: 5_000_000n (5 ADA for script collateral)
 */
const DEFAULT_BUILD_OPTIONS = {
  coinSelection: "largest-first",
  onInsufficientChange: "error",
  setCollateral: 5000000n
};
const makeBuild = (config, programs, options = DEFAULT_BUILD_OPTIONS) => Effect.gen(function* () {
  // Resolve all required resources
  const protocolParameters = yield* resolveProtocolParameters(config, options);
  const changeAddress = yield* resolveChangeAddress(config, options);
  const availableUtxos = yield* resolveAvailableUtxos(config, options);
  // Execute all programs
  yield* Effect.all(programs, {
    concurrency: "unbounded"
  });
  // Run state machine with resolved services
  // Note: FullProtocolParametersTag is provided lazily - evaluation phase will fetch when needed
  const {
    transaction,
    txWithFakeWitnesses
  } = yield* phaseStateMachine.pipe(Effect.provideService(ProtocolParametersTag, protocolParameters), Effect.provideService(ChangeAddressTag, changeAddress), Effect.provideService(AvailableUtxosTag, availableUtxos));
  // Assemble and return final result
  return yield* assembleFinalResult(config, transaction, txWithFakeWitnesses, availableUtxos);
}).pipe(Effect.provideServiceEffect(TxContext, Ref.make(initialTxBuilderState)), Effect.provideService(BuildOptionsTag, {
  ...options,
  evaluator: resolveEvaluator(config, options) ?? options.evaluator,
  slotConfig: resolveSlotConfig(config, options)
}), Effect.provideService(TxBuilderConfigTag, config), Effect.provideServiceEffect(PhaseContextTag, Ref.make({
  phase: "selection",
  attempt: 0,
  calculatedFee: 0n,
  shortfall: 0n,
  changeOutputs: [],
  leftoverAfterFee: CoreAssets.zero,
  canUnfrack: options?.unfrack !== undefined
})));
// Core Effect logic for partial build
const buildPartialEffectCore = (config, programs, _options = DEFAULT_BUILD_OPTIONS) => Effect.gen(function* () {
  // Execute all programs
  yield* Effect.all(programs, {
    concurrency: "unbounded"
  });
  // Return partial transaction (without evaluation)
  return {};
}).pipe(Effect.provideServiceEffect(TxContext, Ref.make(initialTxBuilderState)), Effect.provideService(TxBuilderConfigTag, config), Effect.mapError(error => new TransactionBuilderError({
  message: `Partial build failed: ${error.message}`,
  cause: error
})));
/**
 * Error type for failures in script evaluation.
 *
 * Enhanced with structured failure information including user-provided labels.
 *
 * @since 2.0.0
 * @category errors
 */
export class EvaluationError extends /*#__PURE__*/Data.TaggedError("EvaluationError") {}
/**
 * Context service providing transaction building state to programs.
 * Holds the mutable state Ref - config is passed as a regular parameter.
 *
 * @since 2.0.0
 * @category context
 */
export class TxContext extends /*#__PURE__*/Context.Tag("TxContext")() {}
/**
 * Resolved change address for the current build.
 * This is resolved once at the start of build() from either:
 * - BuildOptions.changeAddress (per-transaction override)
 * - TxBuilderConfig.wallet.Effect.address() (default from wallet)
 *
 * Available to all phase functions via Effect Context.
 *
 * @since 2.0.0
 * @category context
 */
export class ChangeAddressTag extends /*#__PURE__*/Context.Tag("ChangeAddress")() {}
/**
 * Resolved protocol parameters for the current build.
 * This is resolved once at the start of build() from either:
 * - BuildOptions.protocolParameters (per-transaction override)
 * - provider.Effect.getProtocolParameters() (fetched from provider)
 *
 * Available to all phase functions via Effect Context.
 *
 * @since 2.0.0
 * @category context
 */
export class ProtocolParametersTag extends /*#__PURE__*/Context.Tag("ProtocolParameters")() {}
/**
 * Full protocol parameters (including cost models, execution units, etc.) for script evaluation.
 * This is resolved from provider.Effect.getProtocolParameters() and includes all fields
 * needed for UPLC evaluation, unlike the minimal ProtocolParametersTag.
 *
 * Available to evaluation phase via Effect Context.
 *
 * @since 2.0.0
 * @category context
 */
export class FullProtocolParametersTag extends /*#__PURE__*/Context.Tag("FullProtocolParameters")() {}
/**
 * Transaction builder configuration containing provider, wallet, and network information.
 * Available to phases that need to access provider or wallet directly.
 *
 * @since 2.0.0
 * @category context
 */
export class TxBuilderConfigTag extends /*#__PURE__*/Context.Tag("TxBuilderConfig")() {}
/**
 * Resolved available UTxOs for the current build.
 * This is resolved once at the start of build() from either:
 * - BuildOptions.availableUtxos (per-transaction override)
 * - provider.Effect.getUtxos(wallet.address) (default from wallet + provider)
 *
 * Available to all phase functions via Effect Context.
 *
 * @since 2.0.0
 * @category context
 */
export class AvailableUtxosTag extends /*#__PURE__*/Context.Tag("AvailableUtxos")() {}
/**
 * Context tag providing BuildOptions for the current build.
 * Contains build-specific configuration like unfrack, drainTo, onInsufficientChange, etc.
 *
 * @since 2.0.0
 * @category context
 */
export class BuildOptionsTag extends /*#__PURE__*/Context.Tag("BuildOptions")() {}
export function makeTxBuilder(config) {
  const programs = [];
  const txBuilder = {
    // ============================================================================
    // Chainable builder methods - Create ProgramSteps, return same instance
    // ============================================================================
    payToAddress: params => {
      // Create ProgramStep for deferred execution
      const program = createPayToAddressProgram(params);
      programs.push(program);
      return txBuilder; // Return same instance for chaining
    },
    collectFrom: params => {
      // Create ProgramStep for deferred execution
      const program = createCollectFromProgram(params);
      programs.push(program);
      return txBuilder; // Return same instance for chaining
    },
    sendAll: params => {
      // Create ProgramStep for deferred execution
      const program = createSendAllProgram(params);
      programs.push(program);
      return txBuilder; // Return same instance for chaining
    },
    mintAssets: params => {
      // Create ProgramStep for deferred execution
      const program = createMintAssetsProgram(params);
      programs.push(program);
      return txBuilder; // Return same instance for chaining
    },
    readFrom: params => {
      // Create ProgramStep for deferred execution
      const program = createReadFromProgram(params);
      programs.push(program);
      return txBuilder; // Return same instance for chaining
    },
    attachScript: params => {
      // Create ProgramStep for deferred execution
      const program = attachScriptToState(params.script);
      programs.push(program);
      return txBuilder; // Return same instance for chaining
    },
    // Staking/Certificate methods
    registerStake: params => {
      const program = createRegisterStakeProgram(params);
      programs.push(program);
      return txBuilder;
    },
    deregisterStake: params => {
      const program = createDeregisterStakeProgram(params);
      programs.push(program);
      return txBuilder;
    },
    delegateTo: params => {
      const program = createDelegateToProgram(params);
      programs.push(program);
      return txBuilder;
    },
    delegateToPool: params => {
      const program = createDelegateToPoolProgram(params);
      programs.push(program);
      return txBuilder;
    },
    delegateToDRep: params => {
      const program = createDelegateToDRepProgram(params);
      programs.push(program);
      return txBuilder;
    },
    delegateToPoolAndDRep: params => {
      const program = createDelegateToPoolAndDRepProgram(params);
      programs.push(program);
      return txBuilder;
    },
    withdraw: params => {
      const program = createWithdrawProgram(params, config);
      programs.push(program);
      return txBuilder;
    },
    registerAndDelegateTo: params => {
      const program = createRegisterAndDelegateToProgram(params);
      programs.push(program);
      return txBuilder;
    },
    registerDRep: params => {
      const program = createRegisterDRepProgram(params);
      programs.push(program);
      return txBuilder;
    },
    updateDRep: params => {
      const program = createUpdateDRepProgram(params);
      programs.push(program);
      return txBuilder;
    },
    deregisterDRep: params => {
      const program = createDeregisterDRepProgram(params);
      programs.push(program);
      return txBuilder;
    },
    authCommitteeHot: params => {
      const program = createAuthCommitteeHotProgram(params);
      programs.push(program);
      return txBuilder;
    },
    resignCommitteeCold: params => {
      const program = createResignCommitteeColdProgram(params);
      programs.push(program);
      return txBuilder;
    },
    registerPool: params => {
      const program = createRegisterPoolProgram(params);
      programs.push(program);
      return txBuilder;
    },
    retirePool: params => {
      const program = createRetirePoolProgram(params);
      programs.push(program);
      return txBuilder;
    },
    setValidity: params => {
      programs.push(createSetValidityProgram(params));
      return txBuilder;
    },
    vote: params => {
      const program = createVoteProgram(params);
      programs.push(program);
      return txBuilder;
    },
    propose: params => {
      const program = createProposeProgram(params);
      programs.push(program);
      return txBuilder;
    },
    addSigner: params => {
      programs.push(createAddSignerProgram(params));
      return txBuilder;
    },
    attachMetadata: params => {
      programs.push(createAttachMetadataProgram(params));
      return txBuilder;
    },
    compose: other => {
      const otherPrograms = other.getPrograms();
      if (otherPrograms.length > 0) {
        programs.push(...otherPrograms);
      }
      return txBuilder;
    },
    getPrograms: () => [...programs],
    buildEffect: options => {
      return makeBuild(config, programs, options);
    },
    build: options => {
      const effect = makeBuild(config, programs, options);
      return runEffectPromise(options?.debug ? effect.pipe(Effect.provide(Layer.merge(Logger.pretty, Logger.minimumLogLevel(LogLevel.Debug)))) : effect);
    },
    buildEither: options => {
      const effect = makeBuild(config, programs, options).pipe(Effect.either);
      return runEffectPromise(options?.debug ? effect.pipe(Effect.provide(Layer.merge(Logger.pretty, Logger.minimumLogLevel(LogLevel.Debug)))) : effect);
    },
    // ============================================================================
    // Debug methods - Execute with fresh state, return partial transaction
    // ============================================================================
    buildPartialEffect: options => buildPartialEffectCore(config, programs, options),
    buildPartial: options => runEffectPromise(buildPartialEffectCore(config, programs, options))
  };
  return txBuilder;
}
