// @effect/cluster/dist/cjs/EntityAddress.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TypeId = exports.EntityAddressFromSelf = exports.EntityAddress = void 0;
var Equal = _interopRequireWildcard(require("effect/Equal"));
var Hash = _interopRequireWildcard(require("effect/Hash"));
var Schema = _interopRequireWildcard(require("effect/Schema"));
var _EntityId = require("./EntityId.js");
var _EntityType = require("./EntityType.js");
var _ShardId = require("./ShardId.js");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 * @since 1.0.0
 */

const SymbolKey = "@effect/cluster/EntityAddress";
/**
 * @since 1.0.0
 * @category type ids
 */
const TypeId = exports.TypeId = /*#__PURE__*/Symbol.for(SymbolKey);
/**
 * Represents the unique address of an entity within the cluster.
 *
 * @since 1.0.0
 * @category models
 */
class EntityAddress extends /*#__PURE__*/Schema.Class(SymbolKey)({
  shardId: _ShardId.ShardId,
  entityType: _EntityType.EntityType,
  entityId: _EntityId.EntityId
}) {
  /**
   * @since 1.0.0
   */
  [TypeId] = TypeId;
  /**
   * @since 1.0.0
   */
  [Equal.symbol](that) {
    return this.entityType === that.entityType && this.entityId === that.entityId && this.shardId[Equal.symbol](that.shardId);
  }
  /**
   * @since 1.0.0
   */
  [Hash.symbol]() {
    return Hash.cached(this, Hash.string(`${this.entityType}:${this.entityId}:${this.shardId.toString()}`));
  }
}
/**
 * Represents the unique address of an entity within the cluster.
 *
 * @since 1.0.0
 * @category schemas
 */
exports.EntityAddress = EntityAddress;
const EntityAddressFromSelf = exports.EntityAddressFromSelf = /*#__PURE__*/Schema.typeSchema(EntityAddress);
