// @effect/cluster/dist/cjs/Reply.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.serializeLastReceived = exports.serialize = exports.isReply = exports.WithExit = exports.TypeId = exports.ReplyWithContext = exports.Reply = exports.Encoded = exports.Chunk = void 0;
var Rpc = _interopRequireWildcard(require("@effect/rpc/Rpc"));
var Context = _interopRequireWildcard(require("effect/Context"));
var Data = _interopRequireWildcard(require("effect/Data"));
var Effect = _interopRequireWildcard(require("effect/Effect"));
var Exit = _interopRequireWildcard(require("effect/Exit"));
var FiberId = _interopRequireWildcard(require("effect/FiberId"));
var FiberRef = _interopRequireWildcard(require("effect/FiberRef"));
var _Function = require("effect/Function");
var _Predicate = require("effect/Predicate");
var Schema = _interopRequireWildcard(require("effect/Schema"));
var _ClusterError = require("./ClusterError.js");
var _Snowflake = require("./Snowflake.js");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 * @since 1.0.0
 */

/**
 * @since 1.0.0
 * @category type ids
 */
const TypeId = exports.TypeId = /*#__PURE__*/Symbol.for("@effect/cluster/Reply");
/**
 * @since 1.0.0
 * @category guards
 */
const isReply = u => (0, _Predicate.hasProperty)(u, TypeId);
/**
 * @since 1.0.0
 * @category models
 */
exports.isReply = isReply;
class ReplyWithContext extends /*#__PURE__*/Data.TaggedClass("ReplyWithContext") {
  /**
   * @since 1.0.0
   */
  static fromDefect(options) {
    return new ReplyWithContext({
      reply: new WithExit({
        requestId: options.requestId,
        id: options.id,
        exit: Exit.die(Schema.encodeSync(Schema.Defect)(options.defect))
      }),
      context: Context.empty(),
      rpc: neverRpc
    });
  }
  /**
   * @since 1.0.0
   */
  static interrupt(options) {
    return new ReplyWithContext({
      reply: new WithExit({
        requestId: options.requestId,
        id: options.id,
        exit: Exit.interrupt(FiberId.none)
      }),
      context: Context.empty(),
      rpc: neverRpc
    });
  }
}
exports.ReplyWithContext = ReplyWithContext;
const neverRpc = /*#__PURE__*/Rpc.make("Never", {
  success: Schema.Never,
  error: Schema.Never,
  payload: {}
});
const schemaCache = /*#__PURE__*/new WeakMap();
/**
 * @since 1.0.0
 * @category schemas
 */
const Reply = rpc => {
  if (schemaCache.has(rpc)) {
    return schemaCache.get(rpc);
  }
  const schema = Schema.Union(WithExit.schema(rpc), Chunk.schema(rpc));
  schemaCache.set(rpc, schema);
  return schema;
};
/**
 * @since 1.0.0
 * @category schemas
 */
exports.Reply = Reply;
const Encoded = exports.Encoded = /*#__PURE__*/Schema.Union(/*#__PURE__*/Schema.Struct({
  _tag: /*#__PURE__*/Schema.Literal("WithExit"),
  requestId: Schema.String,
  id: Schema.String,
  exit: Schema.Unknown
}), /*#__PURE__*/Schema.Struct({
  _tag: /*#__PURE__*/Schema.Literal("Chunk"),
  requestId: Schema.String,
  id: Schema.String,
  sequence: Schema.Number,
  values: /*#__PURE__*/Schema.Array(Schema.Unknown)
}));
/**
 * @since 1.0.0
 * @category models
 */
class Chunk extends /*#__PURE__*/Data.TaggedClass("Chunk") {
  /**
   * @since 1.0.0
   */
  [TypeId] = TypeId;
  /**
   * @since 1.0.0
   */
  static emptyFrom(requestId) {
    return new Chunk({
      requestId,
      id: (0, _Snowflake.Snowflake)(BigInt(0)),
      sequence: 0,
      values: [undefined]
    });
  }
  /**
   * @since 1.0.0
   */
  static schemaFromSelf = /*#__PURE__*/Schema.declare(u => isReply(u) && u._tag === "Chunk");
  /**
   * @since 1.0.0
   */
  static schema(rpc) {
    const successSchema = rpc.successSchema.success;
    if (!successSchema) {
      return Schema.Never;
    }
    return Schema.transform(Schema.Struct({
      _tag: Schema.Literal("Chunk"),
      requestId: _Snowflake.SnowflakeFromString,
      id: _Snowflake.SnowflakeFromString,
      sequence: Schema.Number,
      values: Schema.NonEmptyArray(successSchema)
    }), Chunk.schemaFromSelf, {
      decode: encoded => new Chunk(encoded),
      encode: _Function.identity
    });
  }
  /**
   * @since 1.0.0
   */
  withRequestId(requestId) {
    return new Chunk({
      ...this,
      requestId
    });
  }
}
/**
 * @since 1.0.0
 * @category models
 */
exports.Chunk = Chunk;
class WithExit extends /*#__PURE__*/Data.TaggedClass("WithExit") {
  /**
   * @since 1.0.0
   */
  [TypeId] = TypeId;
  /**
   * @since 1.0.0
   */
  static schema(rpc) {
    return Schema.transform(Schema.Struct({
      _tag: Schema.Literal("WithExit"),
      requestId: _Snowflake.SnowflakeFromString,
      id: _Snowflake.SnowflakeFromString,
      exit: Rpc.exitSchema(rpc)
    }), Schema.declare(u => isReply(u) && u._tag === "WithExit"), {
      decode: encoded => new WithExit(encoded),
      encode: _Function.identity
    });
  }
  /**
   * @since 1.0.0
   */
  withRequestId(requestId) {
    return new WithExit({
      ...this,
      requestId
    });
  }
}
/**
 * @since 1.0.0
 * @category serialization / deserialization
 */
exports.WithExit = WithExit;
const serialize = self => {
  const schema = Reply(self.rpc);
  return _ClusterError.MalformedMessage.refail(Effect.locally(Schema.encode(schema)(self.reply), FiberRef.currentContext, self.context));
};
/**
 * @since 1.0.0
 * @category serialization / deserialization
 */
exports.serialize = serialize;
const serializeLastReceived = self => {
  if (self.lastReceivedReply._tag === "None") {
    return Effect.succeedNone;
  }
  const schema = Reply(self.rpc);
  return Effect.asSome(_ClusterError.MalformedMessage.refail(Effect.locally(Schema.encode(schema)(self.lastReceivedReply.value), FiberRef.currentContext, self.context)));
};
exports.serializeLastReceived = serializeLastReceived;
