// @effect/cluster/dist/cjs/RunnerAddress.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.make = exports.TypeId = exports.RunnerAddress = void 0;
var Equal = _interopRequireWildcard(require("effect/Equal"));
var Hash = _interopRequireWildcard(require("effect/Hash"));
var _Inspectable = require("effect/Inspectable");
var Schema = _interopRequireWildcard(require("effect/Schema"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 * @since 1.0.0
 */

const SymbolKey = "@effect/cluster/RunnerAddress";
/**
 * @since 1.0.0
 * @category type ids
 */
const TypeId = exports.TypeId = /*#__PURE__*/Symbol.for(SymbolKey);
/**
 * @since 1.0.0
 * @category models
 */
class RunnerAddress extends /*#__PURE__*/Schema.Class(SymbolKey)({
  host: Schema.NonEmptyString,
  port: Schema.Int
}) {
  /**
   * @since 1.0.0
   */
  [TypeId] = TypeId;
  /**
   * @since 1.0.0
   */
  [Equal.symbol](that) {
    return this.host === that.host && this.port === that.port;
  }
  /**
   * @since 1.0.0
   */
  [Hash.symbol]() {
    return Hash.cached(this, Hash.string(this.toString()));
  }
  /**
   * @since 1.0.0
   */
  toString() {
    return `RunnerAddress(${this.host}:${this.port})`;
  }
  /**
   * @since 1.0.0
   */
  [_Inspectable.NodeInspectSymbol]() {
    return this.toString();
  }
}
/**
 * @since 1.0.0
 * @category constructors
 */
exports.RunnerAddress = RunnerAddress;
const make = (host, port) => new RunnerAddress({
  host,
  port
});
exports.make = make;
