// @evolution-sdk/evolution/dist/sdk/provider/internal/Kupo.js
import { Schema as S } from "effect";
export const ValueSchema = /*#__PURE__*/S.Struct({
  coins: S.Number,
  assets: /*#__PURE__*/S.Record({
    key: S.String,
    value: S.Number
  })
});
export const UTxOSchema = /*#__PURE__*/S.Struct({
  transaction_index: S.Number,
  transaction_id: S.String,
  output_index: S.Number,
  address: S.String,
  value: ValueSchema,
  datum_hash: /*#__PURE__*/S.NullOr(S.String),
  datum_type: /*#__PURE__*/S.optional(/*#__PURE__*/S.Literal("hash", "inline")),
  script_hash: /*#__PURE__*/S.NullOr(S.String),
  created_at: /*#__PURE__*/S.Struct({
    slot_no: S.Number,
    header_hash: S.String
  }),
  spent_at: /*#__PURE__*/S.NullOr(/*#__PURE__*/S.Struct({
    slot_no: S.Number,
    header_hash: S.String
  }))
});
export const ScriptSchema = /*#__PURE__*/S.NullOr(/*#__PURE__*/S.Struct({
  language: /*#__PURE__*/S.Literal("native", "plutus:v1", "plutus:v2", "plutus:v3"),
  script: S.String
}));
export const DatumSchema = /*#__PURE__*/S.NullOr(/*#__PURE__*/S.Struct({
  datum: S.String
}));
export const DelegationSchema = /*#__PURE__*/S.NullOr(/*#__PURE__*/S.Record({
  key: S.String,
  value: /*#__PURE__*/S.Struct({
    delegate: /*#__PURE__*/S.Struct({
      id: S.String
    }),
    rewards: /*#__PURE__*/S.Struct({
      ada: /*#__PURE__*/S.Struct({
        lovelace: S.Number
      })
    }),
    deposit: /*#__PURE__*/S.Struct({
      ada: /*#__PURE__*/S.Struct({
        lovelace: S.Number
      })
    })
  })
}));
